{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f5056b6",
   "metadata": {},
   "source": [
    "# Debugging the analysis module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5988d12a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## imports\n",
    "import polars as pl\n",
    "from sklearn.metrics import (\n",
    "    average_precision_score,\n",
    "    balanced_accuracy_score,\n",
    "    confusion_matrix,\n",
    "    f1_score,\n",
    "    roc_auc_score,\n",
    ")\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "# Define a function to compute metrics for each group\n",
    "def compute_aubprc(auprc, prior):\n",
    "    ## calculate the balanced AUPRC based on a prior (incorporating the class imbalance)\n",
    "    return (auprc * (1 - prior)) / ((auprc * (1 - prior)) + ((1 - auprc) * prior))\n",
    "\n",
    "\n",
    "def compute_metrics(group):\n",
    "    ## Calculate all the metrics for evaluating a classifier\n",
    "    y_true = group[\"Label\"].to_numpy()\n",
    "    y_prob = group[\"Prediction\"].to_numpy()\n",
    "    y_pred = (y_prob > 0.5).astype(int)\n",
    "    prior = sum(y_true == 1) / len(y_true)\n",
    "\n",
    "    class_ID = group[\"Classifier_ID\"].unique()[0]\n",
    "\n",
    "    # Compute AUROC\n",
    "    auroc = roc_auc_score(y_true, y_prob)\n",
    "\n",
    "    # Compute AUPRC and balanced AUPRC\n",
    "    auprc = average_precision_score(y_true, y_prob)\n",
    "    aubprc = compute_aubprc(auprc, prior)\n",
    "\n",
    "    # Compute macro-averaged F1 score\n",
    "    macro_f1 = f1_score(y_true, y_pred, average=\"macro\")\n",
    "\n",
    "    # Compute sensitivity and specificity\n",
    "    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()\n",
    "    sensitivity = tp / (tp + fn)\n",
    "    specificity = tn / (tn + fp)\n",
    "\n",
    "    # Compute balanced accuracy\n",
    "    balanced_acc = balanced_accuracy_score(y_true, y_pred)\n",
    "\n",
    "    return {\n",
    "        \"AUROC\": auroc,\n",
    "        \"AUPRC\": auprc,\n",
    "        \"AUBPRC\": aubprc,\n",
    "        \"Macro_F1\": macro_f1,\n",
    "        \"Sensitivity\": sensitivity,\n",
    "        \"Specificity\": specificity,\n",
    "        \"Balanced_Accuracy\": balanced_acc,\n",
    "        \"Classifier_ID\": class_ID,\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "70d38638",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_class_metrics(classifier_info: str, predictions: str, metrics_file: str):\n",
    "    batch_id = [subdir for subdir in classifier_info.split(\"/\") if \"Batch\" in subdir][0]\n",
    "    batch_id = f\"B{batch_id.split('Batch_')[-1]}\"\n",
    "\n",
    "    ## read in classifier info\n",
    "    class_info = pl.read_csv(classifier_info)\n",
    "\n",
    "    ## calculate the class imbalance in training and testing per each classifier\n",
    "    class_info = class_info.with_columns(\n",
    "        (pl.col(\"trainsize_1\") / (pl.col(\"trainsize_0\") + pl.col(\"trainsize_1\"))).alias(\n",
    "            \"train_prob_1\"\n",
    "        ),\n",
    "        (pl.col(\"testsize_1\") / (pl.col(\"testsize_0\") + pl.col(\"testsize_1\"))).alias(\n",
    "            \"test_prob_1\"\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    ## read in predictions\n",
    "    preds = pl.scan_parquet(predictions)\n",
    "    preds = preds.with_columns(pl.lit(batch_id).alias(\"Batch\")).collect()\n",
    "    preds = preds.with_columns(\n",
    "        pl.concat_str(\n",
    "            [pl.col(\"Classifier_ID\"), pl.col(\"Metadata_Protein\"), pl.col(\"Batch\")],\n",
    "            separator=\"_\",\n",
    "        ).alias(\"Full_Classifier_ID\")\n",
    "    )\n",
    "\n",
    "    ## Initialize an empty list to store the results\n",
    "    results = []\n",
    "    classIDs = preds.select(\"Full_Classifier_ID\").to_series().unique().to_list()\n",
    "\n",
    "    # Group by Classifier_ID and compute metrics for each group\n",
    "    for id in tqdm(classIDs):\n",
    "        metrics = compute_metrics(preds.filter(pl.col(\"Full_Classifier_ID\") == id))\n",
    "        metrics[\"Full_Classifier_ID\"] = id\n",
    "        results.append(metrics)\n",
    "\n",
    "    # Convert the results to a Polars DataFrame\n",
    "    metrics_df = pl.DataFrame(results)\n",
    "\n",
    "    # Add classifier info and save\n",
    "    metrics_df = metrics_df.join(class_info, on=\"Classifier_ID\")\n",
    "    metrics_df = metrics_df.with_columns(\n",
    "        (\n",
    "            pl.max_horizontal([\"trainsize_0\", \"trainsize_1\"])\n",
    "            / pl.min_horizontal([\"trainsize_0\", \"trainsize_1\"])\n",
    "        ).alias(\"Training_imbalance\"),\n",
    "        (\n",
    "            pl.max_horizontal([\"testsize_0\", \"testsize_1\"])\n",
    "            / pl.min_horizontal([\"testsize_0\", \"testsize_1\"])\n",
    "        ).alias(\"Testing_imbalance\"),\n",
    "    )\n",
    "    metrics_df.write_csv(metrics_file)\n",
    "    return metrics_df\n",
    "\n",
    "\n",
    "def compute_hits(metrics_file: str, metrics_summary_file: str, trn_imbal_thres: int, min_num_classifier: int):\n",
    "    metrics_df = pl.read_csv(metrics_file)\n",
    "    \n",
    "    batch_id = [subdir for subdir in metrics_file.split(\"/\") if \"Batch\" in subdir][0]\n",
    "    batch_id = f\"B{batch_id.split('Batch_')[-1]}\"\n",
    "\n",
    "    # Add useful columns (type, batch)\n",
    "    metrics_df = metrics_df.with_columns(\n",
    "        #  Group the same alleles from platemap together, by extracting the substring that:\n",
    "        #  1. Has a digit (\\d) immediately before it (anchors the match at a number)\n",
    "        #  2. Starts with 'A' and then as few characters as needed (A.*?), captured as group 1\n",
    "        #  3. Stops right before the literal 'T'\n",
    "        pl.col(\"Plate\").str.extract(r\"\\d(A.*?)T\", 1).alias(\"Allele_set\"),\n",
    "    )\n",
    "\n",
    "    metrics_df = metrics_df.with_columns(\n",
    "        pl.when(pl.col(\"Full_Classifier_ID\").str.contains(\"true\"))\n",
    "        .then(pl.lit(\"localization\"))\n",
    "        .otherwise(pl.lit(\"morphology\"))\n",
    "        .alias(\"Classifier_type\"),\n",
    "        pl.col(\"Full_Classifier_ID\").str.split(\"_\").list.last().alias(\"Batch\"),\n",
    "    )\n",
    "\n",
    "    # Filter based on class imbalance\n",
    "    metrics_ctrl = (\n",
    "        metrics_df.filter(\n",
    "            (pl.col(\"Training_imbalance\") < trn_imbal_thres) & (pl.col(\"Metadata_Control\"))\n",
    "        )\n",
    "        .select([\"Classifier_type\", \"Batch\", \"AUROC\"])\n",
    "        .group_by([\"Classifier_type\", \"Batch\"])\n",
    "        .quantile(0.99)\n",
    "    ).rename({\"AUROC\": \"AUROC_thresh\"})\n",
    "\n",
    "    # Merge with metrics_df and decide whether it passed the threshold\n",
    "    metrics_df = metrics_df.join(metrics_ctrl, on=[\"Classifier_type\", \"Batch\"])\n",
    "\n",
    "    # Must be at least min_class_num classifiers per batch\n",
    "    # Number of classifiers is the same for localization and morph, so just use morph\n",
    "    classifier_count = (\n",
    "        metrics_df.filter(\n",
    "            (~pl.col(\"Metadata_Control\"))\n",
    "            & (pl.col(\"Classifier_type\") == \"localization\")\n",
    "        )\n",
    "        .group_by([\"allele_0\", \"Allele_set\", \"Batch\", \"allele_1\"])\n",
    "        .agg([pl.len().alias(\"Number_classifiers\")])\n",
    "    )\n",
    "    classifier_count = classifier_count.pivot(\n",
    "        index=[\"allele_0\", \"allele_1\", \"Allele_set\"],\n",
    "        on=\"Batch\",\n",
    "        values=\"Number_classifiers\",\n",
    "    )\n",
    "    print(\"Total number of unique classifiers:\", classifier_count.shape[0])\n",
    "    print(\"Total number of unique variant alleles:\", len(classifier_count.select(\"allele_0\").to_series().unique().to_list()))\n",
    "    print(\"Total number of unique WT genes:\", len(classifier_count.select(\"allele_1\").to_series().unique().to_list()))\n",
    "    print(\"==========================================================================\")\n",
    "\n",
    "    # Must be at least min_class_num classifiers per batch\n",
    "    # Number of classifiers is the same for localization and morph, so just use morph\n",
    "    classifier_count = (\n",
    "        metrics_df.filter(\n",
    "            (pl.col(\"Training_imbalance\") < trn_imbal_thres)\n",
    "            & (~pl.col(\"Metadata_Control\"))\n",
    "            & (pl.col(\"Classifier_type\") == \"localization\")\n",
    "        )\n",
    "        .group_by([\"allele_0\", \"Allele_set\", \"Batch\", \"allele_1\"])\n",
    "        .agg([pl.len().alias(\"Number_classifiers\")])\n",
    "    )\n",
    "    classifier_count = classifier_count.pivot(\n",
    "        index=[\"allele_0\", \"allele_1\", \"Allele_set\"],\n",
    "        on=\"Batch\",\n",
    "        values=\"Number_classifiers\",\n",
    "    )\n",
    "    print(\"After filtering out classifiers with training imbalance > 3:\")\n",
    "    print(\"Total number of unique classifiers:\", classifier_count.shape)\n",
    "    print(\"Total number of unique variant alleles:\", len(classifier_count.select(\"allele_0\").to_series().unique().to_list()))\n",
    "    print(\"Total number of unique WT genes:\", len(classifier_count.select(\"allele_1\").to_series().unique().to_list()))\n",
    "    print(\"==========================================================================\")\n",
    "\n",
    "    classifier_count = classifier_count.filter(\n",
    "        (pl.col(batch_id) >= min_num_classifier)\n",
    "    )\n",
    "    print(\"After filtering out alleles with available number of classifiers < 2:\")\n",
    "    print(\"Total number of unique classifiers:\", classifier_count.shape)\n",
    "    print(\"Total number of unique variant alleles:\", len(classifier_count.select(\"allele_0\").to_series().unique().to_list()))\n",
    "    print(\"Total number of unique WT genes:\", len(classifier_count.select(\"allele_1\").to_series().unique().to_list()))\n",
    "\n",
    "    # filter based on this\n",
    "    keep_alleles = classifier_count.select(\"allele_0\").to_series().unique().to_list()\n",
    "    metrics_df = metrics_df.filter(\n",
    "        ~((~pl.col(\"Metadata_Control\")) & ~pl.col(\"allele_0\").is_in(keep_alleles))\n",
    "    )\n",
    "\n",
    "    # Filter by imbalance and calculate mean AUROC for each batch\n",
    "    metrics_wtvar = (\n",
    "        (\n",
    "            metrics_df.filter(\n",
    "                (~pl.col(\"Metadata_Control\")) ## pl.col(\"Training_imbalance\") < trn_imbal_thres) & \n",
    "            )\n",
    "        )\n",
    "        .select([\n",
    "            \"AUROC\",\n",
    "            \"Classifier_type\",\n",
    "            \"Batch\",\n",
    "            \"AUROC_thresh\",\n",
    "            \"allele_0\",\n",
    "            \"trainsize_0\",\n",
    "            \"testsize_0\",\n",
    "            \"trainsize_1\",\n",
    "            \"testsize_1\",\n",
    "            \"Allele_set\",\n",
    "            \"Training_imbalance\",\n",
    "        ])\n",
    "        .group_by([\"Classifier_type\", \"allele_0\", \"Allele_set\", \"Batch\", \"AUROC_thresh\"])\n",
    "        .agg([\n",
    "            pl.all()\n",
    "            .exclude([\"Classifier_type\", \"allele_0\", \"Allele_set\", \"Batch\", \"AUROC_thresh\"])\n",
    "            .mean()\n",
    "            .name.suffix(\"_mean\")\n",
    "        ])\n",
    "    )\n",
    "    return metrics_wtvar\n",
    "    \n",
    "    # Write out results\n",
    "    # metrics_wtvar.write_csv(metrics_summary_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ff33d8c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique classifiers: 439\n",
      "Total number of unique variant alleles: 439\n",
      "Total number of unique WT genes: 23\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (385, 4)\n",
      "Total number of unique variant alleles: 385\n",
      "Total number of unique WT genes: 23\n",
      "==========================================================================\n",
      "After filtering out alleles with available number of classifiers < 2:\n",
      "Total number of unique classifiers: (360, 4)\n",
      "Total number of unique variant alleles: 360\n",
      "Total number of unique WT genes: 23\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (720, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Classifier_type</th><th>allele_0</th><th>Allele_set</th><th>Batch</th><th>AUROC_thresh</th><th>AUROC_mean</th><th>trainsize_0_mean</th><th>testsize_0_mean</th><th>trainsize_1_mean</th><th>testsize_1_mean</th><th>Training_imbalance_mean</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;localization&quot;</td><td>&quot;KRAS_Gly12Asp&quot;</td><td>&quot;A7A8P1_&quot;</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.75201</td><td>1269.0</td><td>423.0</td><td>1374.0</td><td>458.0</td><td>1.147595</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;CCM2_Gly407Asp&quot;</td><td>&quot;A7A8P2_&quot;</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.783035</td><td>1341.75</td><td>447.25</td><td>3903.0</td><td>1301.0</td><td>2.915331</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;F9_Gly442Arg&quot;</td><td>&quot;A7A8P1_&quot;</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.549513</td><td>1122.0</td><td>374.0</td><td>757.5</td><td>252.5</td><td>1.484793</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;RET_Val262Ala&quot;</td><td>&quot;A7A8P2_&quot;</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.835261</td><td>870.75</td><td>290.25</td><td>1044.0</td><td>348.0</td><td>1.206279</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;BRAF_Gly596Val&quot;</td><td>&quot;A7A8P1_&quot;</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.893082</td><td>1570.5</td><td>523.5</td><td>1419.0</td><td>473.0</td><td>1.101321</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;BRCA1_Ala102Gly&quot;</td><td>&quot;A7A8P1_&quot;</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.877495</td><td>1113.75</td><td>371.25</td><td>1239.0</td><td>413.0</td><td>1.170862</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;BRCA1_Thr587Arg&quot;</td><td>&quot;A7A8P1_&quot;</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.905352</td><td>1232.25</td><td>410.75</td><td>1239.0</td><td>413.0</td><td>1.145495</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;RPS19_Arg62Trp&quot;</td><td>&quot;A7A8P2_&quot;</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.725234</td><td>420.75</td><td>140.25</td><td>442.5</td><td>147.5</td><td>1.075846</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;SDHD_His102Pro&quot;</td><td>&quot;A7A8P2_&quot;</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.623399</td><td>1325.25</td><td>441.75</td><td>1876.5</td><td>625.5</td><td>1.418962</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;CCM2_Phe91Ile&quot;</td><td>&quot;A7A8P2_&quot;</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.711015</td><td>1708.5</td><td>569.5</td><td>3903.0</td><td>1301.0</td><td>2.304615</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (720, 11)\n",
       "┌────────────┬────────────┬────────────┬───────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ Classifier ┆ allele_0   ┆ Allele_set ┆ Batch ┆ … ┆ testsize_ ┆ trainsize ┆ testsize_ ┆ Training_ │\n",
       "│ _type      ┆ ---        ┆ ---        ┆ ---   ┆   ┆ 0_mean    ┆ _1_mean   ┆ 1_mean    ┆ imbalance │\n",
       "│ ---        ┆ str        ┆ str        ┆ str   ┆   ┆ ---       ┆ ---       ┆ ---       ┆ _mean     │\n",
       "│ str        ┆            ┆            ┆       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ ---       │\n",
       "│            ┆            ┆            ┆       ┆   ┆           ┆           ┆           ┆ f64       │\n",
       "╞════════════╪════════════╪════════════╪═══════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ localizati ┆ KRAS_Gly12 ┆ A7A8P1_    ┆ B13   ┆ … ┆ 423.0     ┆ 1374.0    ┆ 458.0     ┆ 1.147595  │\n",
       "│ on         ┆ Asp        ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ CCM2_Gly40 ┆ A7A8P2_    ┆ B13   ┆ … ┆ 447.25    ┆ 3903.0    ┆ 1301.0    ┆ 2.915331  │\n",
       "│ on         ┆ 7Asp       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ F9_Gly442A ┆ A7A8P1_    ┆ B13   ┆ … ┆ 374.0     ┆ 757.5     ┆ 252.5     ┆ 1.484793  │\n",
       "│ on         ┆ rg         ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ RET_Val262 ┆ A7A8P2_    ┆ B13   ┆ … ┆ 290.25    ┆ 1044.0    ┆ 348.0     ┆ 1.206279  │\n",
       "│            ┆ Ala        ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ BRAF_Gly59 ┆ A7A8P1_    ┆ B13   ┆ … ┆ 523.5     ┆ 1419.0    ┆ 473.0     ┆ 1.101321  │\n",
       "│ on         ┆ 6Val       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ …          ┆ …          ┆ …          ┆ …     ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │\n",
       "│ morphology ┆ BRCA1_Ala1 ┆ A7A8P1_    ┆ B13   ┆ … ┆ 371.25    ┆ 1239.0    ┆ 413.0     ┆ 1.170862  │\n",
       "│            ┆ 02Gly      ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ BRCA1_Thr5 ┆ A7A8P1_    ┆ B13   ┆ … ┆ 410.75    ┆ 1239.0    ┆ 413.0     ┆ 1.145495  │\n",
       "│            ┆ 87Arg      ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ RPS19_Arg6 ┆ A7A8P2_    ┆ B13   ┆ … ┆ 140.25    ┆ 442.5     ┆ 147.5     ┆ 1.075846  │\n",
       "│            ┆ 2Trp       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ SDHD_His10 ┆ A7A8P2_    ┆ B13   ┆ … ┆ 441.75    ┆ 1876.5    ┆ 625.5     ┆ 1.418962  │\n",
       "│ on         ┆ 2Pro       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ CCM2_Phe91 ┆ A7A8P2_    ┆ B13   ┆ … ┆ 569.5     ┆ 3903.0    ┆ 1301.0    ┆ 2.304615  │\n",
       "│            ┆ Ile        ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "└────────────┴────────────┴────────────┴───────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metrics_file = \"../outputs/classification_analyses/2025_01_27_Batch_13/profiles_tcdropped_filtered_var_mad_outlier_featselect/metrics.csv\"\n",
    "compute_hits(metrics_file, \"\", trn_imbal_thres=3, min_num_classifier=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e45b3efd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (720, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Classifier_type</th><th>allele_0</th><th>Allele_set</th><th>Batch</th><th>AUROC_thresh</th><th>AUROC_mean</th><th>trainsize_0_mean</th><th>testsize_0_mean</th><th>trainsize_1_mean</th><th>testsize_1_mean</th><th>Training_imbalance_mean</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;morphology&quot;</td><td>&quot;F9_Arg3His&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.619266</td><td>554.25</td><td>184.75</td><td>757.5</td><td>252.5</td><td>1.366262</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;CCM2_Gly5Ser&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.76576</td><td>1458.75</td><td>486.25</td><td>3903.0</td><td>1301.0</td><td>2.684427</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;CTCF_Asp46Asn&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.79788</td><td>363.75</td><td>121.25</td><td>720.75</td><td>240.25</td><td>2.013674</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;BRAF_Gln257Arg&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.709575</td><td>810.0</td><td>270.0</td><td>1419.0</td><td>473.0</td><td>1.843124</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;RAD51D_Val56Gly&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.696832</td><td>1219.5</td><td>406.5</td><td>984.0</td><td>328.0</td><td>1.237085</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;CCM2_Arg412Gln&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.653131</td><td>2172.75</td><td>724.25</td><td>3903.0</td><td>1301.0</td><td>1.822314</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;F9_Ile7Phe&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.626554</td><td>598.5</td><td>199.5</td><td>757.5</td><td>252.5</td><td>1.268307</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;KRAS_Gly12Asp&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.786289</td><td>1269.0</td><td>423.0</td><td>1374.0</td><td>458.0</td><td>1.147595</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;F9_Glu79Asp&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.923142</td><td>0.583162</td><td>494.25</td><td>164.75</td><td>757.5</td><td>252.5</td><td>1.528803</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;SDHD_His102Leu&quot;</td><td>null</td><td>&quot;B13&quot;</td><td>0.937407</td><td>0.827895</td><td>2185.5</td><td>728.5</td><td>1876.5</td><td>625.5</td><td>1.165004</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (720, 11)\n",
       "┌────────────┬────────────┬────────────┬───────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ Classifier ┆ allele_0   ┆ Allele_set ┆ Batch ┆ … ┆ testsize_ ┆ trainsize ┆ testsize_ ┆ Training_ │\n",
       "│ _type      ┆ ---        ┆ ---        ┆ ---   ┆   ┆ 0_mean    ┆ _1_mean   ┆ 1_mean    ┆ imbalance │\n",
       "│ ---        ┆ str        ┆ str        ┆ str   ┆   ┆ ---       ┆ ---       ┆ ---       ┆ _mean     │\n",
       "│ str        ┆            ┆            ┆       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ ---       │\n",
       "│            ┆            ┆            ┆       ┆   ┆           ┆           ┆           ┆ f64       │\n",
       "╞════════════╪════════════╪════════════╪═══════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ morphology ┆ F9_Arg3His ┆ null       ┆ B13   ┆ … ┆ 184.75    ┆ 757.5     ┆ 252.5     ┆ 1.366262  │\n",
       "│ morphology ┆ CCM2_Gly5S ┆ null       ┆ B13   ┆ … ┆ 486.25    ┆ 3903.0    ┆ 1301.0    ┆ 2.684427  │\n",
       "│            ┆ er         ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ CTCF_Asp46 ┆ null       ┆ B13   ┆ … ┆ 121.25    ┆ 720.75    ┆ 240.25    ┆ 2.013674  │\n",
       "│            ┆ Asn        ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ BRAF_Gln25 ┆ null       ┆ B13   ┆ … ┆ 270.0     ┆ 1419.0    ┆ 473.0     ┆ 1.843124  │\n",
       "│ on         ┆ 7Arg       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ RAD51D_Val ┆ null       ┆ B13   ┆ … ┆ 406.5     ┆ 984.0     ┆ 328.0     ┆ 1.237085  │\n",
       "│ on         ┆ 56Gly      ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ …          ┆ …          ┆ …          ┆ …     ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │\n",
       "│ morphology ┆ CCM2_Arg41 ┆ null       ┆ B13   ┆ … ┆ 724.25    ┆ 3903.0    ┆ 1301.0    ┆ 1.822314  │\n",
       "│            ┆ 2Gln       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ F9_Ile7Phe ┆ null       ┆ B13   ┆ … ┆ 199.5     ┆ 757.5     ┆ 252.5     ┆ 1.268307  │\n",
       "│ morphology ┆ KRAS_Gly12 ┆ null       ┆ B13   ┆ … ┆ 423.0     ┆ 1374.0    ┆ 458.0     ┆ 1.147595  │\n",
       "│            ┆ Asp        ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ F9_Glu79As ┆ null       ┆ B13   ┆ … ┆ 164.75    ┆ 757.5     ┆ 252.5     ┆ 1.528803  │\n",
       "│ on         ┆ p          ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ SDHD_His10 ┆ null       ┆ B13   ┆ … ┆ 728.5     ┆ 1876.5    ┆ 625.5     ┆ 1.165004  │\n",
       "│            ┆ 2Leu       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "└────────────┴────────────┴────────────┴───────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metrics_file = \"../outputs/classification_analyses/2025_01_27_Batch_13/profiles_tcdropped_filtered_var_mad_outlier_featselect/metrics_summary.csv\"\n",
    "metrics_summ = pl.read_csv(metrics_file)\n",
    "metrics_summ"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varchamp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
