{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "419c9d9d",
   "metadata": {},
   "source": [
    "# Debugging the classify module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8fc2eab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Classification pipeline\"\"\"\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "from itertools import combinations\n",
    "from typing import Union\n",
    "import cupy as cp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "import xgboost as xgb\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from tqdm import tqdm\n",
    "from tqdm.contrib.concurrent import thread_map\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sys.path.append(\"..\")\n",
    "from utils import find_feat_cols, find_meta_cols, remove_nan_infs_columns\n",
    "\n",
    "\n",
    "def classifier(df_train, df_test, log_file, target=\"Label\", shuffle=False):\n",
    "    \"\"\"\n",
    "    This function runs classification.\n",
    "    \"\"\"\n",
    "\n",
    "    feat_col = find_feat_cols(df_train)\n",
    "    feat_col.remove(target)\n",
    "\n",
    "    x_train, y_train = cp.array(df_train[feat_col].to_numpy()), df_train[[target]]\n",
    "    x_test, y_test = cp.array(df_test[feat_col].to_numpy()), df_test[[target]]\n",
    "\n",
    "    num_pos = df_train[df_train[target] == 1].shape[0]\n",
    "    num_neg = df_train[df_train[target] == 0].shape[0]\n",
    "\n",
    "    if (num_pos == 0) or (num_neg == 0):\n",
    "        log_file.write(f\"Missing positive/negative labels in {df_train['Metadata_Plate'].unique()}, {df_train['Metadata_symbol'].unique()} wells: {df_train['Metadata_well_position'].unique()}\")\n",
    "        log_file.write(f\"Size of pos: {num_pos}, Size of neg: {num_neg}\\n\")\n",
    "\n",
    "        print(f\"size of pos: {num_pos}, size of neg: {num_neg}\")\n",
    "        feat_importances = pd.Series(np.nan, index=df_train[feat_col].columns)\n",
    "        return feat_importances, np.nan\n",
    "\n",
    "    scale_pos_weight = num_neg / num_pos\n",
    "\n",
    "    if (scale_pos_weight > 100) or (scale_pos_weight < 0.01):\n",
    "        log_file.write(f\"Extreme class imbalance in {df_train['Metadata_Plate'].unique()}, {df_train['Metadata_symbol'].unique()} wells: {df_train['Metadata_well_position'].unique()}\")\n",
    "        log_file.write(f\"Scale_pos_weight: {scale_pos_weight}, Size of pos: {num_pos}, Size of neg: {num_neg}\\n\")\n",
    "        print(\n",
    "            f\"scale_pos_weight: {scale_pos_weight}, size of pos: {num_pos}, size of neg: {num_neg}\"\n",
    "        )\n",
    "        feat_importances = pd.Series(np.nan, index=df_train[feat_col].columns)\n",
    "        return feat_importances, np.nan\n",
    "\n",
    "    le = LabelEncoder()\n",
    "    y_train = cp.array(le.fit_transform(y_train))\n",
    "    y_test = cp.array(le.fit_transform(y_test))\n",
    "\n",
    "    if shuffle:\n",
    "        # Create shuffled train labels\n",
    "        y_train_shuff = y_train.copy()\n",
    "        y_train_shuff[\"Label\"] = np.random.permutation(y_train.values)\n",
    "\n",
    "    model = xgb.XGBClassifier(\n",
    "        objective=\"binary:logistic\",\n",
    "        n_estimators=150,\n",
    "        tree_method=\"hist\",\n",
    "        device=\"cuda\",\n",
    "        learning_rate=0.05,\n",
    "        scale_pos_weight=scale_pos_weight,\n",
    "    ).fit(x_train, y_train, verbose=False)\n",
    "\n",
    "    # get predictions and scores\n",
    "    pred_score = model.predict_proba(x_test)[:, 1]\n",
    "\n",
    "    # Return classifier info\n",
    "    info_0 = df_test[df_test[\"Label\"] == 0].iloc[0]\n",
    "    info_1 = df_test[df_test[\"Label\"] == 1].iloc[0]\n",
    "    class_ID = (\n",
    "        info_0[\"Metadata_Plate\"]\n",
    "        + \"_\"\n",
    "        + info_0[\"Metadata_well_position\"]\n",
    "        + \"_\"\n",
    "        + info_1[\"Metadata_well_position\"]\n",
    "    )\n",
    "    classifier_df = pd.DataFrame({\n",
    "        \"Classifier_ID\": [class_ID],\n",
    "        \"Plate\": [info_0[\"Metadata_Plate\"]],\n",
    "        \"trainsize_0\": [sum(y_train.get() == 0)],\n",
    "        \"testsize_0\": [sum(y_test.get() == 0)],\n",
    "        \"well_0\": [info_0[\"Metadata_well_position\"]],\n",
    "        \"allele_0\": [info_0[\"Metadata_gene_allele\"]],\n",
    "        \"trainsize_1\": [sum(y_train.get() == 1)],\n",
    "        \"testsize_1\": [sum(y_test.get() == 1)],\n",
    "        \"well_1\": [info_1[\"Metadata_well_position\"]],\n",
    "        \"allele_1\": [info_1[\"Metadata_gene_allele\"]],\n",
    "    })\n",
    "\n",
    "    # Store feature importance\n",
    "    feat_importances = pd.Series(\n",
    "        model.feature_importances_, index=df_train[feat_col].columns\n",
    "    )\n",
    "\n",
    "    # Return cell-level predictions\n",
    "    cellID = df_test.apply(\n",
    "        lambda row: f\"{row['Metadata_Plate']}_{row['Metadata_well_position']}_{row['Metadata_ImageNumber']}_{row['Metadata_ObjectNumber']}\",\n",
    "        axis=1,\n",
    "    ).to_list()\n",
    "\n",
    "    pred_df = pd.DataFrame({\n",
    "        \"Classifier_ID\": class_ID,\n",
    "        \"CellID\": cellID,\n",
    "        \"Label\": y_test.get(),\n",
    "        \"Prediction\": pred_score,\n",
    "    })\n",
    "\n",
    "    return feat_importances, classifier_df, pred_df\n",
    "\n",
    "\n",
    "def get_classifier_features(dframe: pd.DataFrame, protein_feat: bool):\n",
    "    \"\"\"Helper function to get dframe containing protein or non-protein features\"\"\"\n",
    "    feat_col = find_feat_cols(dframe)\n",
    "    meta_col = find_meta_cols(dframe)\n",
    "\n",
    "    if protein_feat:\n",
    "        feat_col = [\n",
    "            i\n",
    "            for i in feat_col\n",
    "            if (\"GFP\" in i)\n",
    "            and (\"DNA\" not in i)\n",
    "            and (\"AGP\" not in i)\n",
    "            and (\"Mito\" not in i)\n",
    "            and (\"Brightfield\" not in i)\n",
    "        ]\n",
    "    else:\n",
    "        feat_col = [\n",
    "            i\n",
    "            for i in feat_col\n",
    "            if (\"GFP\" not in i) and (\"Brightfield\" not in i) ## and (\"AGP\" not in i), test without AGP features\n",
    "        ]\n",
    "\n",
    "    dframe = pd.concat([dframe[meta_col], dframe[feat_col]], axis=1)\n",
    "    return dframe\n",
    "\n",
    "\n",
    "def stratify_by_plate(df_sampled: pd.DataFrame, plate: str):\n",
    "    \"\"\"Stratify dframe by plate\"\"\"\n",
    "    # print(df_sampled.head())\n",
    "    df_sampled_platemap = plate.split(\"_T\")[0]\n",
    "    platemaps = df_sampled[df_sampled[\"Metadata_Plate\"].str.contains(df_sampled_platemap)][\"Metadata_plate_map_name\"].to_list()\n",
    "    assert(len(set(platemaps))==1), \"Only one platemap should be associated with plate: {plate}.\"\n",
    "    platemap = platemaps[0]\n",
    "\n",
    "    # Train on data from same platemap but other plates\n",
    "    df_train = df_sampled[\n",
    "        (df_sampled[\"Metadata_plate_map_name\"] == platemap)\n",
    "        & (df_sampled[\"Metadata_Plate\"] != plate)\n",
    "    ].reset_index(drop=True)\n",
    "\n",
    "    df_test = df_sampled[df_sampled[\"Metadata_Plate\"] == plate].reset_index(drop=True)\n",
    "    # print(df_train.head())\n",
    "    # print(df_test.head())\n",
    "\n",
    "    return df_train, df_test\n",
    "\n",
    "\n",
    "def experimental_runner(\n",
    "    exp_dframe: pd.DataFrame,\n",
    "    pq_writer,\n",
    "    log_file,\n",
    "    protein=True,\n",
    "    group_key_one=\"Metadata_symbol\",\n",
    "    group_key_two=\"Metadata_gene_allele\",\n",
    "    threshold_key=\"Metadata_node_type\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Run Reference v.s. Variant experiments\n",
    "    \"\"\"\n",
    "    exp_dframe = get_classifier_features(exp_dframe, protein)\n",
    "    feat_cols = find_feat_cols(exp_dframe)\n",
    "    feat_cols = [i for i in feat_cols if i != \"Label\"]\n",
    "\n",
    "    group_list = []\n",
    "    pair_list = []\n",
    "    feat_list = []\n",
    "    info_list = []\n",
    "\n",
    "    log_file.write(f\"Running XGBboost classifiers w/ protein {protein} on target variants:\\n\")\n",
    "    groups = exp_dframe.groupby(group_key_one).groups\n",
    "    for key in tqdm(groups.keys()):\n",
    "        dframe_grouped = exp_dframe.loc[groups[key]].reset_index(drop=True)\n",
    "\n",
    "        # Ensure this gene has both reference and variants\n",
    "        if dframe_grouped[threshold_key].unique().size < 2:\n",
    "            continue\n",
    "\n",
    "        df_group_one = dframe_grouped[\n",
    "            dframe_grouped[\"Metadata_node_type\"] == \"disease_wt\"\n",
    "        ].reset_index(drop=True)\n",
    "        df_group_one[\"Label\"] = 1\n",
    "\n",
    "        subgroups = (\n",
    "            dframe_grouped[dframe_grouped[\"Metadata_node_type\"] == \"allele\"]\n",
    "            .groupby(group_key_two)\n",
    "            .groups\n",
    "        )\n",
    "\n",
    "        for subkey in subgroups.keys():\n",
    "            df_group_two = dframe_grouped.loc[subgroups[subkey]].reset_index(drop=True)\n",
    "            df_group_two[\"Label\"] = 0\n",
    "\n",
    "            df_sampled = pd.concat([df_group_one, df_group_two], ignore_index=True)\n",
    "            plate_list = get_common_plates(df_group_one, df_group_two)\n",
    "\n",
    "            def classify_by_plate_helper(plate):\n",
    "                df_train, df_test = stratify_by_plate(df_sampled, plate)\n",
    "                feat_importances, classifier_info, predictions = classifier(\n",
    "                    df_train, df_test, log_file\n",
    "                )\n",
    "                print(feat_importances)\n",
    "                return {plate: [feat_importances, classifier_info, predictions]}\n",
    "\n",
    "            # try run classifier\n",
    "            try:\n",
    "                result = thread_map(classify_by_plate_helper, plate_list)\n",
    "\n",
    "                pred_list = []\n",
    "                for res in result:\n",
    "                    if len(list(res.values())[0]) == 3:\n",
    "                        feat_list.append(list(res.values())[0][0])\n",
    "                        group_list.append(key)\n",
    "                        pair_list.append(f\"{key}_{subkey}\")\n",
    "                        info_list.append(list(res.values())[0][1])\n",
    "                        pred_list.append(list(res.values())[0][2])\n",
    "                    else:\n",
    "                        print(\"res length not 3!\")\n",
    "                        feat_list.append([None] * len(feat_cols))\n",
    "                        group_list.append(key)\n",
    "                        pair_list.append(f\"{key}_{subkey}\")\n",
    "                        info_list.append([None] * 10)\n",
    "\n",
    "                cell_preds = pd.concat(pred_list, axis=0)\n",
    "                cell_preds[\"Metadata_Protein\"] = protein\n",
    "                cell_preds[\"Metadata_Control\"] = False\n",
    "                table = pa.Table.from_pandas(cell_preds, preserve_index=False)\n",
    "                pq_writer.write_table(table)\n",
    "            except Exception as e:\n",
    "                print(e)\n",
    "                log_file.write(f\"{key}, {subkey} error: {e}\\n\")\n",
    "            break\n",
    "        break\n",
    "\n",
    "    # Store feature importance\n",
    "    df_feat_one = pd.DataFrame({\"Group1\": group_list, \"Group2\": pair_list})\n",
    "    df_feat_two = pd.DataFrame(feat_list)\n",
    "    df_feat = pd.concat([df_feat_one, df_feat_two], axis=1)\n",
    "    df_feat[\"Metadata_Protein\"] = protein\n",
    "    df_feat[\"Metadata_Control\"] = False\n",
    "\n",
    "    # process classifier info\n",
    "    df_result = pd.concat(info_list, ignore_index=True)\n",
    "    df_result[\"Metadata_Control\"] = False\n",
    "\n",
    "    log_file.write(f\"Finished running XGBboost classifiers w/ protein {protein} on target variants.\\n\")\n",
    "    log_file.write(f\"===========================================================================\\n\\n\")\n",
    "    return df_feat, df_result\n",
    "\n",
    "\n",
    "def get_common_plates(dframe1, dframe2):\n",
    "    \"\"\"Helper func: get common plates in two dataframes\"\"\"\n",
    "    plate_list = list(\n",
    "        set(list(dframe1[\"Metadata_Plate\"].unique()))\n",
    "        & set(list(dframe2[\"Metadata_Plate\"].unique()))\n",
    "    )\n",
    "    return plate_list\n",
    "\n",
    "\n",
    "def control_group_runner(\n",
    "    ctrl_dframe: pd.DataFrame,\n",
    "    pq_writer,\n",
    "    log_file,\n",
    "    group_key_one=\"Metadata_gene_allele\",\n",
    "    group_key_two=\"Metadata_plate_map_name\",\n",
    "    group_key_three=\"Metadata_well_position\",\n",
    "    threshold_key=\"Metadata_well_position\",\n",
    "    protein=True,\n",
    "):\n",
    "    \"\"\"\n",
    "    Run null control experiments.\n",
    "    \"\"\"\n",
    "    ctrl_dframe = get_classifier_features(ctrl_dframe, protein)\n",
    "    feat_cols = find_feat_cols(ctrl_dframe)\n",
    "    feat_cols = [i for i in feat_cols if i != \"Label\"]\n",
    "\n",
    "    group_list = []\n",
    "    pair_list = []\n",
    "    feat_list = []\n",
    "    info_list = []\n",
    "\n",
    "    log_file.write(f\"Running XGBboost classifiers w/ protein {protein} on control alleles:\\n\")\n",
    "    groups = ctrl_dframe.groupby(group_key_one).groups\n",
    "    for key in tqdm(groups.keys()):\n",
    "        # groupby alleles\n",
    "        dframe_grouped = ctrl_dframe.loc[groups[key]].reset_index(drop=True)\n",
    "\n",
    "        # Skip controls with no replicates\n",
    "        if dframe_grouped[threshold_key].unique().size < 2:\n",
    "            continue\n",
    "\n",
    "        # group by platemap\n",
    "        subgroups = dframe_grouped.groupby(group_key_two).groups\n",
    "        for key_two in subgroups.keys():\n",
    "            dframe_grouped_two = dframe_grouped.loc[subgroups[key_two]].reset_index(\n",
    "                drop=True\n",
    "            )\n",
    "            # If a well is not present on all four plates, drop well\n",
    "            well_count = dframe_grouped_two.groupby([\"Metadata_Well\"])[\n",
    "                \"Metadata_Plate\"\n",
    "            ].nunique()\n",
    "            well_to_drop = well_count[well_count < 4].index\n",
    "            dframe_grouped_two = dframe_grouped_two[\n",
    "                ~dframe_grouped_two[\"Metadata_Well\"].isin(well_to_drop)\n",
    "            ].reset_index(drop=True)\n",
    "\n",
    "            # group by well\n",
    "            sub_sub_groups = dframe_grouped_two.groupby(group_key_three).groups\n",
    "            sampled_pairs = list(combinations(list(sub_sub_groups.keys()), r=2))\n",
    "\n",
    "            for idx1, idx2 in sampled_pairs:\n",
    "                df_group_one = dframe_grouped_two.loc[sub_sub_groups[idx1]].reset_index(\n",
    "                    drop=True\n",
    "                )\n",
    "                df_group_one[\"Label\"] = 1\n",
    "                df_group_two = dframe_grouped_two.loc[sub_sub_groups[idx2]].reset_index(\n",
    "                    drop=True\n",
    "                )\n",
    "                df_group_two[\"Label\"] = 0\n",
    "                df_sampled = pd.concat([df_group_one, df_group_two], ignore_index=True)\n",
    "\n",
    "                try:\n",
    "                    plate_list = get_common_plates(df_group_one, df_group_two)\n",
    "\n",
    "                    def classify_by_plate_helper(plate):\n",
    "                        df_train, df_test = stratify_by_plate(df_sampled, plate)\n",
    "                        feat_importances, classifier_info, predictions = classifier(\n",
    "                            df_train, df_test, log_file\n",
    "                        )\n",
    "                        print(feat_importances)\n",
    "                        return {plate: [feat_importances, classifier_info, predictions]}\n",
    "\n",
    "                    result = thread_map(classify_by_plate_helper, plate_list)\n",
    "\n",
    "                    pred_list = []\n",
    "                    for res in result:\n",
    "                        if len(list(res.values())[0]) == 3:\n",
    "                            feat_list.append(list(res.values())[0][0])\n",
    "                            group_list.append(key)\n",
    "                            pair_list.append(f\"{idx1}_{idx2}\")\n",
    "                            info_list.append(list(res.values())[0][1])\n",
    "                            pred_list.append(list(res.values())[0][2])\n",
    "                        else:\n",
    "                            print(\"res length does not equal three!\")\n",
    "                            feat_list.append([None] * len(feat_cols))\n",
    "                            group_list.append(key)\n",
    "                            pair_list.append(f\"{idx1}_{idx2}\")\n",
    "                            info_list.append([None] * 10)\n",
    "\n",
    "                    cell_preds = pd.concat(pred_list, axis=0)\n",
    "                    cell_preds[\"Metadata_Protein\"] = protein\n",
    "                    cell_preds[\"Metadata_Control\"] = True\n",
    "                    table = pa.Table.from_pandas(cell_preds, preserve_index=False)\n",
    "                    pq_writer.write_table(table)\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "                    log_file.write(f\"{key}, {key_two} error: {e}, wells per ctrl: {sub_sub_groups}\\n\")\n",
    "                break\n",
    "            break\n",
    "\n",
    "    # Store feature importance\n",
    "    df_feat_one = pd.DataFrame({\"Group1\": group_list, \"Group2\": pair_list})\n",
    "    df_feat_two = pd.DataFrame(feat_list)\n",
    "    df_feat = pd.concat([df_feat_one, df_feat_two], axis=1)\n",
    "    df_feat[\"Metadata_Protein\"] = protein\n",
    "    df_feat[\"Metadata_Control\"] = True\n",
    "\n",
    "    # process classifier info\n",
    "    df_result = pd.concat(info_list, ignore_index=True)\n",
    "    df_result[\"Metadata_Control\"] = True\n",
    "\n",
    "    log_file.write(f\"Finished running XGBboost classifiers w/ protein {protein} on control alleles.\\n\")\n",
    "    log_file.write(f\"===========================================================================\\n\\n\")\n",
    "    return df_feat, df_result\n",
    "\n",
    "\n",
    "def control_type_helper(col_annot: str):\n",
    "    \"\"\"helper func for annotating column \"Metadata_control\" \"\"\"\n",
    "    ## Only TC, NC, PC are used for constructing the null distribution because of multiple duplicates \n",
    "    if col_annot in [\"TC\", \"NC\", \"PC\"]:\n",
    "        return True\n",
    "    ## else labeled as not controls\n",
    "    elif col_annot in [\"disease_wt\", \"allele\", \"cPC\", \"cNC\"]:\n",
    "        return False\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def add_control_annot(dframe):\n",
    "    \"\"\"annotating column \"Metadata_control\" \"\"\"\n",
    "    if \"Metadata_control\" not in dframe.columns:\n",
    "        dframe[\"Metadata_control\"] = dframe[\"Metadata_node_type\"].apply(\n",
    "            lambda x: control_type_helper(x)\n",
    "        )\n",
    "    return dframe\n",
    "\n",
    "\n",
    "def drop_low_cc_wells(dframe, cc_thresh, log_file):\n",
    "    # Drop wells with cell counts lower than the threshold\n",
    "    dframe[\"Metadata_Cell_ID\"] = dframe.index\n",
    "    cell_count = (\n",
    "        dframe.groupby([\"Metadata_Plate\", \"Metadata_Well\"])[\"Metadata_Cell_ID\"]\n",
    "        .count()\n",
    "        .reset_index(name=\"Metadata_Cell_Count\")\n",
    "    )\n",
    "    ## get the cell counts per well per plate\n",
    "    dframe = dframe.merge(\n",
    "        cell_count,\n",
    "        on=[\"Metadata_Plate\", \"Metadata_Well\"],\n",
    "    )\n",
    "    dframe_dropped = (\n",
    "        dframe[dframe[\"Metadata_Cell_Count\"] < cc_thresh]\n",
    "    )\n",
    "    ## keep track of the alleles in a log file\n",
    "    log_file.write(f\"Number of wells dropped due to cell counts < {cc_thresh}: {len((dframe_dropped['Metadata_Plate']+dframe_dropped['Metadata_Well']+dframe_dropped['Metadata_gene_allele']).unique())}\\n\")\n",
    "    dframe_dropped = dframe_dropped.drop_duplicates(subset=[\"Metadata_Plate\", \"Metadata_Well\"])\n",
    "    if (dframe_dropped.shape[0] > 0):\n",
    "        for idx in dframe_dropped.index:\n",
    "            log_file.write(f\"{dframe_dropped.loc[idx, 'Metadata_Plate']}, {dframe_dropped.loc[idx, 'Metadata_Well']}:{dframe_dropped.loc[idx, 'Metadata_gene_allele']}\\n\")\n",
    "            # print(f\"{dframe_dropped.loc[idx, 'Metadata_Plate']}, {dframe_dropped.loc[idx, 'Metadata_Well']}:{dframe_dropped.loc[idx, 'Metadata_gene_allele']}\\n\")\n",
    "    ## keep only the wells with cc >= cc_thresh\n",
    "    dframe = (\n",
    "        dframe[dframe[\"Metadata_Cell_Count\"] >= cc_thresh]\n",
    "        .drop(columns=[\"Metadata_Cell_Count\"])\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "    return dframe\n",
    "\n",
    "\n",
    "def run_classify_workflow(\n",
    "    input_path: str,\n",
    "    feat_output_path: str,\n",
    "    info_output_path: str,\n",
    "    preds_output_path: str,\n",
    "    cc_threshold: int,\n",
    "    use_gpu: Union[str, None] = \"0,1\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Run workflow for single-cell classification\n",
    "    \"\"\"\n",
    "    if use_gpu is not None:\n",
    "        os.environ[\"CUDA_VISIBLE_DEVICES\"] = use_gpu\n",
    "\n",
    "    # Initialize parquet for cell-level predictions\n",
    "    if os.path.exists(preds_output_path):\n",
    "        os.remove(preds_output_path)\n",
    "\n",
    "    logfile_path = os.path.join(*[preds_output_path.split(\"/\")[:-1], \"classify.log\"])\n",
    "\n",
    "    schema = pa.schema([\n",
    "        (\"Classifier_ID\", pa.string()),\n",
    "        (\"CellID\", pa.string()),\n",
    "        (\"Label\", pa.int64()),\n",
    "        (\"Prediction\", pa.float32()),\n",
    "        (\"Metadata_Protein\", pa.bool_()),\n",
    "        (\"Metadata_Control\", pa.bool_()),\n",
    "    ])\n",
    "    writer = pq.ParquetWriter(preds_output_path, schema, compression=\"gzip\")\n",
    "\n",
    "    # Add CellID column\n",
    "    dframe = (\n",
    "        pl.scan_parquet(input_path)\n",
    "        .with_columns(\n",
    "            pl.concat_str(\n",
    "                [\n",
    "                    \"Metadata_Plate\",\n",
    "                    \"Metadata_well_position\",\n",
    "                    \"Metadata_ImageNumber\",\n",
    "                    \"Metadata_ObjectNumber\",\n",
    "                ],\n",
    "                separator=\"_\",\n",
    "            ).alias(\"Metadata_CellID\")\n",
    "        )\n",
    "        .collect()\n",
    "        .to_pandas()\n",
    "    )\n",
    "\n",
    "    feat_col = find_feat_cols(dframe)\n",
    "\n",
    "    try:\n",
    "        assert (\n",
    "            ~np.isnan(dframe[feat_col]).any().any()\n",
    "        ), \"Dataframe contains no NaN features.\"\n",
    "        assert (\n",
    "            np.isfinite(dframe[feat_col]).all().all()\n",
    "        ), \"Dataframe contains finite feature values.\"\n",
    "    except AssertionError:\n",
    "        dframe = remove_nan_infs_columns(dframe)\n",
    "\n",
    "    # Filter rows with NaN Metadata\n",
    "    dframe = dframe[~dframe[\"Metadata_well_position\"].isna()]\n",
    "    dframe = add_control_annot(dframe)\n",
    "    dframe = dframe[~dframe[\"Metadata_control\"].isna()]\n",
    "\n",
    "    # Split data into controls and alleles\n",
    "    df_exp = dframe[~dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(drop=True)\n",
    "    df_control = dframe[dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(\n",
    "        drop=True\n",
    "    )\n",
    "\n",
    "    # Remove any remaining TC from analysis\n",
    "    df_control = df_control[df_control[\"Metadata_node_type\"] != \"TC\"].reset_index(\n",
    "        drop=True\n",
    "    )\n",
    "\n",
    "    with open(logfile_path, \"w\") as log_file:\n",
    "        log_file.write(f\"===============================================================================================================================================================\\n\")\n",
    "        log_file.write(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "        print(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "\n",
    "        df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "        log_file.write(\"Dropping low cell count wells in ref. vs variant alleles:\\n\")\n",
    "        print(\"Dropping low cell count wells in ref. vs variant alleles:\")\n",
    "        df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "        log_file.write(f\"===============================================================================================================================================================\\n\\n\")\n",
    "\n",
    "        # Filter out wells with fewer than the cell count threhsold\n",
    "        df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "        df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "\n",
    "        # Protein feature analysis\n",
    "        df_feat_pro_con, df_result_pro_con = control_group_runner(\n",
    "            df_control, pq_writer=writer, log_file=log_file, protein=True\n",
    "        )\n",
    "        df_feat_pro_exp, df_result_pro_exp = experimental_runner(\n",
    "            df_exp, pq_writer=writer, log_file=log_file, protein=True\n",
    "        )\n",
    "\n",
    "        # Non-protein feature analysis\n",
    "        df_feat_no_pro_con, df_result_no_pro_con = control_group_runner(\n",
    "            df_control, pq_writer=writer, log_file=log_file, protein=False\n",
    "        )\n",
    "        df_feat_no_pro_exp, df_result_no_pro_exp = experimental_runner(\n",
    "            df_exp, pq_writer=writer, log_file=log_file, protein=False\n",
    "        )\n",
    "        writer.close()\n",
    "\n",
    "    # Concatenate results for both protein and non-protein\n",
    "    df_feat = pd.concat(\n",
    "        [df_feat_pro_con, df_feat_no_pro_con, df_feat_pro_exp, df_feat_no_pro_exp],\n",
    "        ignore_index=True,\n",
    "    )\n",
    "    df_result = pd.concat(\n",
    "        [\n",
    "            df_result_pro_con,\n",
    "            df_result_no_pro_con,\n",
    "            df_result_pro_exp,\n",
    "            df_result_no_pro_exp,\n",
    "        ],\n",
    "        ignore_index=True,\n",
    "    )\n",
    "    df_result = df_result.drop_duplicates()\n",
    "\n",
    "    # Write out feature importance and classifier info\n",
    "    df_feat.to_csv(feat_output_path, index=False)\n",
    "    df_result.to_csv(info_output_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f245a91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = \"../outputs/batch_profiles/2025_01_27_Batch_13/profiles_tcdropped_filtered_var_mad_outlier_featselect.parquet\"\n",
    "# Add CellID column\n",
    "dframe = (\n",
    "    pl.scan_parquet(input_path)\n",
    "    .with_columns(\n",
    "        pl.concat_str(\n",
    "            [\n",
    "                \"Metadata_Plate\",\n",
    "                \"Metadata_well_position\",\n",
    "                \"Metadata_ImageNumber\",\n",
    "                \"Metadata_ObjectNumber\",\n",
    "            ],\n",
    "            separator=\"_\",\n",
    "        ).alias(\"Metadata_CellID\")\n",
    "    )\n",
    "    .collect()\n",
    "    .to_pandas()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3134f0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_plate_map_name</th>\n",
       "      <th>Metadata_well_position</th>\n",
       "      <th>Metadata_symbol</th>\n",
       "      <th>Metadata_gene_allele</th>\n",
       "      <th>Metadata_imaging_well</th>\n",
       "      <th>Metadata_imaging_plate_R1</th>\n",
       "      <th>Metadata_imaging_plate_R2</th>\n",
       "      <th>Metadata_node_type</th>\n",
       "      <th>Metadata_orf_id_wt</th>\n",
       "      <th>Metadata_ccsb_mutation_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Cells_RadialDistribution_FracAtD_mito_tubeness_7of20</th>\n",
       "      <th>Nuclei_ObjectSkeleton_TotalObjectSkeletonLength_mito_skel</th>\n",
       "      <th>Cytoplasm_RadialDistribution_RadialCV_DNA_4of10</th>\n",
       "      <th>Cells_RadialDistribution_RadialCV_GFP_3of10</th>\n",
       "      <th>Nuclei_RadialDistribution_FracAtD_DNA_6of10</th>\n",
       "      <th>Nuclei_Texture_Contrast_AGP_20_00_256</th>\n",
       "      <th>Nuclei_Texture_SumAverage_GFP_5_01_256</th>\n",
       "      <th>Cytoplasm_Texture_InfoMeas1_Mito_20_02_256</th>\n",
       "      <th>Cytoplasm_Granularity_1_AGP</th>\n",
       "      <th>Metadata_CellID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>A01</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>ABCD1</td>\n",
       "      <td>A01</td>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B14A7A8P1_R2</td>\n",
       "      <td>disease_wt</td>\n",
       "      <td>100008637.0</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.436310</td>\n",
       "      <td>3.047979</td>\n",
       "      <td>-1.856935</td>\n",
       "      <td>-1.315689</td>\n",
       "      <td>0.010491</td>\n",
       "      <td>1.847285</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.303491</td>\n",
       "      <td>-0.248480</td>\n",
       "      <td>2025_01_27_B13A7A8P1_T1_A01_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B17</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>KRAS_Phe156Ile</td>\n",
       "      <td>B17</td>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B14A7A8P1_R2</td>\n",
       "      <td>allele</td>\n",
       "      <td>3084.0</td>\n",
       "      <td>CCSBVarC006102</td>\n",
       "      <td>...</td>\n",
       "      <td>2.299274</td>\n",
       "      <td>-2.124484</td>\n",
       "      <td>-0.361802</td>\n",
       "      <td>-1.398442</td>\n",
       "      <td>-0.930505</td>\n",
       "      <td>-1.377401</td>\n",
       "      <td>1.416060</td>\n",
       "      <td>-6.973239</td>\n",
       "      <td>2.271974</td>\n",
       "      <td>2025_01_27_B13A7A8P1_T1_B17_366_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B17</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>KRAS_Phe156Ile</td>\n",
       "      <td>B17</td>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B14A7A8P1_R2</td>\n",
       "      <td>allele</td>\n",
       "      <td>3084.0</td>\n",
       "      <td>CCSBVarC006102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.961759</td>\n",
       "      <td>0.609714</td>\n",
       "      <td>-0.928406</td>\n",
       "      <td>0.711069</td>\n",
       "      <td>-2.359273</td>\n",
       "      <td>0.014116</td>\n",
       "      <td>0.633440</td>\n",
       "      <td>-2.078081</td>\n",
       "      <td>0.869858</td>\n",
       "      <td>2025_01_27_B13A7A8P1_T1_B17_365_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B17</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>KRAS_Phe156Ile</td>\n",
       "      <td>B17</td>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B14A7A8P1_R2</td>\n",
       "      <td>allele</td>\n",
       "      <td>3084.0</td>\n",
       "      <td>CCSBVarC006102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.529808</td>\n",
       "      <td>1.467804</td>\n",
       "      <td>-1.945400</td>\n",
       "      <td>-0.888838</td>\n",
       "      <td>-0.206016</td>\n",
       "      <td>1.751616</td>\n",
       "      <td>1.212706</td>\n",
       "      <td>1.674929</td>\n",
       "      <td>0.653823</td>\n",
       "      <td>2025_01_27_B13A7A8P1_T1_B17_364_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B17</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>KRAS_Phe156Ile</td>\n",
       "      <td>B17</td>\n",
       "      <td>B13A7A8P1_R1</td>\n",
       "      <td>B14A7A8P1_R2</td>\n",
       "      <td>allele</td>\n",
       "      <td>3084.0</td>\n",
       "      <td>CCSBVarC006102</td>\n",
       "      <td>...</td>\n",
       "      <td>4.751731</td>\n",
       "      <td>0.255296</td>\n",
       "      <td>-1.097230</td>\n",
       "      <td>0.382016</td>\n",
       "      <td>-0.751108</td>\n",
       "      <td>0.296083</td>\n",
       "      <td>6.045101</td>\n",
       "      <td>0.306296</td>\n",
       "      <td>-0.375300</td>\n",
       "      <td>2025_01_27_B13A7A8P1_T1_B17_363_33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1163 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_plate_map_name Metadata_well_position Metadata_symbol  \\\n",
       "0            B13A7A8P1_R1                    A01           ABCD1   \n",
       "1            B13A7A8P1_R1                    B17            KRAS   \n",
       "2            B13A7A8P1_R1                    B17            KRAS   \n",
       "3            B13A7A8P1_R1                    B17            KRAS   \n",
       "4            B13A7A8P1_R1                    B17            KRAS   \n",
       "\n",
       "  Metadata_gene_allele Metadata_imaging_well Metadata_imaging_plate_R1  \\\n",
       "0                ABCD1                   A01              B13A7A8P1_R1   \n",
       "1       KRAS_Phe156Ile                   B17              B13A7A8P1_R1   \n",
       "2       KRAS_Phe156Ile                   B17              B13A7A8P1_R1   \n",
       "3       KRAS_Phe156Ile                   B17              B13A7A8P1_R1   \n",
       "4       KRAS_Phe156Ile                   B17              B13A7A8P1_R1   \n",
       "\n",
       "  Metadata_imaging_plate_R2 Metadata_node_type  Metadata_orf_id_wt  \\\n",
       "0              B14A7A8P1_R2         disease_wt         100008637.0   \n",
       "1              B14A7A8P1_R2             allele              3084.0   \n",
       "2              B14A7A8P1_R2             allele              3084.0   \n",
       "3              B14A7A8P1_R2             allele              3084.0   \n",
       "4              B14A7A8P1_R2             allele              3084.0   \n",
       "\n",
       "  Metadata_ccsb_mutation_id  ...  \\\n",
       "0                      None  ...   \n",
       "1            CCSBVarC006102  ...   \n",
       "2            CCSBVarC006102  ...   \n",
       "3            CCSBVarC006102  ...   \n",
       "4            CCSBVarC006102  ...   \n",
       "\n",
       "  Cells_RadialDistribution_FracAtD_mito_tubeness_7of20  \\\n",
       "0                                          -0.436310     \n",
       "1                                           2.299274     \n",
       "2                                           0.961759     \n",
       "3                                           0.529808     \n",
       "4                                           4.751731     \n",
       "\n",
       "  Nuclei_ObjectSkeleton_TotalObjectSkeletonLength_mito_skel  \\\n",
       "0                                           3.047979          \n",
       "1                                          -2.124484          \n",
       "2                                           0.609714          \n",
       "3                                           1.467804          \n",
       "4                                           0.255296          \n",
       "\n",
       "   Cytoplasm_RadialDistribution_RadialCV_DNA_4of10  \\\n",
       "0                                        -1.856935   \n",
       "1                                        -0.361802   \n",
       "2                                        -0.928406   \n",
       "3                                        -1.945400   \n",
       "4                                        -1.097230   \n",
       "\n",
       "   Cells_RadialDistribution_RadialCV_GFP_3of10  \\\n",
       "0                                    -1.315689   \n",
       "1                                    -1.398442   \n",
       "2                                     0.711069   \n",
       "3                                    -0.888838   \n",
       "4                                     0.382016   \n",
       "\n",
       "   Nuclei_RadialDistribution_FracAtD_DNA_6of10  \\\n",
       "0                                     0.010491   \n",
       "1                                    -0.930505   \n",
       "2                                    -2.359273   \n",
       "3                                    -0.206016   \n",
       "4                                    -0.751108   \n",
       "\n",
       "   Nuclei_Texture_Contrast_AGP_20_00_256  \\\n",
       "0                               1.847285   \n",
       "1                              -1.377401   \n",
       "2                               0.014116   \n",
       "3                               1.751616   \n",
       "4                               0.296083   \n",
       "\n",
       "   Nuclei_Texture_SumAverage_GFP_5_01_256  \\\n",
       "0                               -1.000000   \n",
       "1                                1.416060   \n",
       "2                                0.633440   \n",
       "3                                1.212706   \n",
       "4                                6.045101   \n",
       "\n",
       "   Cytoplasm_Texture_InfoMeas1_Mito_20_02_256  Cytoplasm_Granularity_1_AGP  \\\n",
       "0                                    0.303491                    -0.248480   \n",
       "1                                   -6.973239                     2.271974   \n",
       "2                                   -2.078081                     0.869858   \n",
       "3                                    1.674929                     0.653823   \n",
       "4                                    0.306296                    -0.375300   \n",
       "\n",
       "                      Metadata_CellID  \n",
       "0     2025_01_27_B13A7A8P1_T1_A01_1_1  \n",
       "1  2025_01_27_B13A7A8P1_T1_B17_366_33  \n",
       "2  2025_01_27_B13A7A8P1_T1_B17_365_33  \n",
       "3  2025_01_27_B13A7A8P1_T1_B17_364_33  \n",
       "4  2025_01_27_B13A7A8P1_T1_B17_363_33  \n",
       "\n",
       "[5 rows x 1163 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(dframe.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bc41d58f",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_col = find_feat_cols(dframe)\n",
    "# feat_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8a6ec66f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping low cell count wells in control alleles:\n",
      "\n",
      "Dropping low cell count wells in ref. vs variant alleles:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/6 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d41fb0867a4e4fa6ba3647ba4d2dc243",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.007782\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004222\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.005065\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.009622\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005279\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.004163\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004804\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.005009\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006681\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.003119\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.003713\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.003973\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.005678\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006389\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005712\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.004398\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007507\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.004624\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006352\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.002927\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.003503\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004202\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.003526\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.059139\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005853\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.005575\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004341\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.004726\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.008664\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.004935\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.014482\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004185\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.005354\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.020882\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005392\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.004727\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.009466\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.004188\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006508\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.003500\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 17%|█▋        | 1/6 [00:02<00:10,  2.19s/it]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8cb4ff4b9d4a44058bab1738693914ae",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.002434\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.013310\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004992\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.007294\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.007900\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007436\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.006560\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.007256\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005319\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.000000\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 33%|███▎      | 2/6 [00:03<00:05,  1.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.001790\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.003807\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.006521\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006267\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006603\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007996\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005392\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006991\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.004636\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.005626\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.005765\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.010377\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004754\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.016476\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006372\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.010501\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.008949\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.007633\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005993\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.001317\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.003377\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.007673\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.003729\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.007671\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.008469\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.009151\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005409\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006105\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005324\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.000000\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c95fab811efb4e629f083284260a6dcd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.006459\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.006254\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004451\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006484\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006155\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007896\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005988\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.005557\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006450\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.009572\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.004199\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004179\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.006176\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006736\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006762\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.010186\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.006187\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006281\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005379\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.017379\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.003113\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.003864\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004569\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.005454\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.004826\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.010536\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005922\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.005505\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.004800\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.006162\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.004277\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004214\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.005592\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006549\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005751\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.008643\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005761\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.005074\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005630\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.005073\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 50%|█████     | 3/6 [00:04<00:04,  1.48s/it]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "798162c10bf84a559d5ad7e290d7137b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 67%|██████▋   | 4/6 [00:05<00:02,  1.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.010462\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004435\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004505\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.005268\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.004713\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007216\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004082\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.007626\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.004191\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.009534\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.004025\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.006344\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.008956\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.025307\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006913\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.008640\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007410\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006283\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006041\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.001385\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.002399\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004708\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.006737\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.016565\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.004640\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.006113\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007697\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006609\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.004771\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.013719\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.005676\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.006289\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.007249\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.004363\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.007947\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.005092\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007323\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.003322\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005849\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.006997\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f2e2ed5e87fe488d9c186ec75d2ad783",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.004311\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.005444\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.003726\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006266\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005575\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.005444\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004617\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.008498\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.007300\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.009945\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.006232\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.004630\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.007755\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006346\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.007161\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007856\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004018\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006211\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006325\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.007665\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.005771\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.007706\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.007547\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.013258\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.003392\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.008349\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005125\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.007131\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.007256\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.009289\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 83%|████████▎ | 5/6 [00:06<00:01,  1.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.004644\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.005721\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004657\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006512\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005669\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.004246\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004522\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.008141\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006467\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.011291\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c398d21dbd9e4e7083101392533516b2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.007421\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.005047\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004987\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.005432\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005569\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007537\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007560\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.004495\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.009224\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.005468\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.006449\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.005087\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.006176\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006436\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.005750\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.008423\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.006686\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.005600\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.008190\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.007563\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.006893\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.007738\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004872\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006949\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006747\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.008345\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007150\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006413\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005301\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.006025\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:07<00:00,  1.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.005638\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.005203\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.007684\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.006051\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006461\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007556\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007058\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.005599\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005497\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.005311\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|          | 0/33 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "58d03f1ab4ff47bf9eaa777b6099db6e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.003090\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.005610\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.007223\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.011327\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.006032\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007233\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.008353\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.007664\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.003296\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.004548\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.017626\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.007202\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.005994\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.003133\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.004049\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.007383\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.004869\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.008672\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.006485\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.012359\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.010099\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.006399\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.004460\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.010617\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.004362\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.004601\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.005308\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.006146\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.004013\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.005987\n",
      "Length: 155, dtype: float32\n",
      "Cytoplasm_Texture_InfoMeas1_GFP_5_02_256        0.006024\n",
      "Nuclei_Texture_Correlation_GFP_5_00_256         0.006639\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_8of10    0.005456\n",
      "Cytoplasm_Intensity_MaxIntensity_GFP            0.009477\n",
      "Nuclei_RadialDistribution_FracAtD_GFP_7of10     0.003564\n",
      "                                                  ...   \n",
      "Cells_Intensity_MinIntensity_GFP                0.008067\n",
      "Cells_RadialDistribution_MeanFrac_GFP_8of10     0.007280\n",
      "Nuclei_RadialDistribution_RadialCV_GFP_2of10    0.004135\n",
      "Cells_RadialDistribution_RadialCV_GFP_3of10     0.005138\n",
      "Nuclei_Texture_SumAverage_GFP_5_01_256          0.000425\n",
      "Length: 155, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/33 [00:00<?, ?it/s]\n",
      "  0%|          | 0/6 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e0fb09040b7a4186b9db450b03df9c4f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000892\n",
      "Nuclei_Granularity_6_Mito                          0.001205\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.003281\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000983\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000383\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000606\n",
      "Cytoplasm_Granularity_1_AGP                        0.000000\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000018\n",
      "Nuclei_Granularity_6_Mito                          0.000000\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000374\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000732\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000868\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000770\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.001332\n",
      "Nuclei_Granularity_6_Mito                          0.001263\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000968\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000565\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000777\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000467\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 17%|█▋        | 1/6 [00:01<00:06,  1.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000000\n",
      "Nuclei_Granularity_6_Mito                          0.002168\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.002350\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000645\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000968\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000239\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.002554\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000165\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.003357\n",
      "Cytoplasm_Granularity_1_AGP                        0.000389\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "58d9069a06ac4a9e885c12905fff8eab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000406\n",
      "Nuclei_Granularity_6_Mito                          0.000446\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.002953\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000432\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000909\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000284\n",
      "Cytoplasm_Granularity_1_AGP                        0.001779\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000899\n",
      "Nuclei_Granularity_6_Mito                          0.003377\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000264\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.006858\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.008663\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000269\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000759\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000786\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000497\n",
      "Nuclei_Granularity_6_Mito                          0.000421\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.006388\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.001150\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000520\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000106\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000864\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 33%|███▎      | 2/6 [00:02<00:04,  1.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.002276\n",
      "Nuclei_Granularity_6_Mito                          0.000000\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.004581\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000698\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.001555\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.001267\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000340\n",
      "Cytoplasm_Granularity_1_AGP                        0.000000\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8329649ebd1947b6bcbb8235ab7c779d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.006202\n",
      "Nuclei_Granularity_6_Mito                          0.000604\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.001186\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000272\n",
      "Cytoplasm_Granularity_1_AGP                        0.007582\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.001213\n",
      "Nuclei_Granularity_6_Mito                          0.004593\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.001990\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000384\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000109\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.001488\n",
      "Cytoplasm_Granularity_1_AGP                        0.005069\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000000\n",
      "Nuclei_Granularity_6_Mito                          0.000000\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.001116\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000877\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.047834\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 50%|█████     | 3/6 [00:04<00:04,  1.57s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000719\n",
      "Nuclei_Granularity_6_Mito                          0.001684\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000128\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000793\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.014151\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c8e85b48f51e49bb9f57daca8bdeb6a0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 67%|██████▋   | 4/6 [00:05<00:02,  1.43s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000857\n",
      "Nuclei_Granularity_6_Mito                          0.000386\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.007968\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.003158\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.001307\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000776\n",
      "Cytoplasm_Granularity_1_AGP                        0.000423\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000526\n",
      "Nuclei_Granularity_6_Mito                          0.000000\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000243\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.001466\n",
      "Nuclei_Granularity_6_Mito                          0.000000\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000771\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000504\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.004469\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.003320\n",
      "Cytoplasm_Granularity_1_AGP                        0.000000\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          5.324361e-07\n",
      "Nuclei_Granularity_6_Mito                          0.000000e+00\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          7.376888e-06\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        9.278784e-04\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000e+00\n",
      "                                                       ...     \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000e+00\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        8.836617e-04\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              1.964445e-03\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000e+00\n",
      "Cytoplasm_Granularity_1_AGP                        0.000000e+00\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4641b358e3d1491b85a08f8cdfd3bff7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 83%|████████▎ | 5/6 [00:07<00:01,  1.41s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.001442\n",
      "Nuclei_Granularity_6_Mito                          0.002554\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000653\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.001330\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.005366\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000692\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.001962\n",
      "Cytoplasm_Granularity_1_AGP                        0.006508\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.002932\n",
      "Nuclei_Granularity_6_Mito                          0.001139\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.001376\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000956\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000605\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000266\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.002251\n",
      "Cytoplasm_Granularity_1_AGP                        0.003488\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.001349\n",
      "Nuclei_Granularity_6_Mito                          0.000605\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.001449\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000367\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.002610\n",
      "Cytoplasm_Granularity_1_AGP                        0.000000\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000514\n",
      "Nuclei_Granularity_6_Mito                          0.001466\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.004394\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.001111\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000582\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.009695\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3c003e9708e34936a8b73ef6ea2a5518",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:08<00:00,  1.40s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000677\n",
      "Nuclei_Granularity_6_Mito                          0.001304\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.001113\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000668\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.002260\n",
      "Cytoplasm_Granularity_1_AGP                        0.001012\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.003186\n",
      "Nuclei_Granularity_6_Mito                          0.001167\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.008263\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000337\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000000\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.003888\n",
      "Cytoplasm_Granularity_1_AGP                        0.002371\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.001450\n",
      "Nuclei_Granularity_6_Mito                          0.000061\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000000\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.003135\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.001272\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.004540\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000161\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.002345\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.002142\n",
      "Nuclei_Granularity_6_Mito                          0.000445\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.002740\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.001619\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.001973\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000678\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000592\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000690\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.001244\n",
      "Cytoplasm_Granularity_1_AGP                        0.001293\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|          | 0/33 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "14f2794a63d24b7394907c35a62db79f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclei_AreaShape_Solidity                          0.000580\n",
      "Nuclei_Granularity_6_Mito                          0.000893\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.001732\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.002101\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000578\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000512\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000848\n",
      "Nuclei_Granularity_6_Mito                          0.001614\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000128\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000000\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000218\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000000\n",
      "Cytoplasm_Granularity_1_AGP                        0.000000\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000000\n",
      "Nuclei_Granularity_6_Mito                          0.001163\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.004664\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000000\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.001296\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000780\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.006151\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.000821\n",
      "Cytoplasm_Granularity_1_AGP                        0.000717\n",
      "Length: 935, dtype: float32\n",
      "Nuclei_AreaShape_Solidity                          0.000424\n",
      "Nuclei_Granularity_6_Mito                          0.003123\n",
      "Nuclei_Texture_SumVariance_Mito_20_03_256          0.000937\n",
      "Cells_RadialDistribution_RadialCV_AGP_7of10        0.000933\n",
      "Nuclei_Texture_InfoMeas2_Mito_5_00_256             0.000172\n",
      "                                                     ...   \n",
      "Cytoplasm_RadialDistribution_RadialCV_DNA_4of10    0.000373\n",
      "Nuclei_RadialDistribution_FracAtD_DNA_6of10        0.000000\n",
      "Nuclei_Texture_Contrast_AGP_20_00_256              0.000000\n",
      "Cytoplasm_Texture_InfoMeas1_Mito_20_02_256         0.002359\n",
      "Cytoplasm_Granularity_1_AGP                        0.000211\n",
      "Length: 935, dtype: float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/33 [00:01<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "# Initialize parquet for cell-level predictions\n",
    "batch = \"2025_01_27_Batch_13\"\n",
    "pipeline = \"profiles_tcdropped_filtered_var_mad_outlier_featselect\"\n",
    "preds_output_path = f\"../outputs/results/{batch}/{pipeline}/predictions.parquet\"\n",
    "cc_threshold = 20\n",
    "\n",
    "if os.path.exists(preds_output_path):\n",
    "    os.remove(preds_output_path)\n",
    "\n",
    "if not os.path.exists(f\"../outputs/results/{batch}/{pipeline}\"):\n",
    "    os.makedirs(f\"../outputs/results/{batch}/{pipeline}\")\n",
    "\n",
    "logfile_path = os.path.join('/'.join(preds_output_path.split(\"/\")[:-1]), \"classify.log\")\n",
    "# logfile_path\n",
    "\n",
    "schema = pa.schema([\n",
    "    (\"Classifier_ID\", pa.string()),\n",
    "    (\"CellID\", pa.string()),\n",
    "    (\"Label\", pa.int64()),\n",
    "    (\"Prediction\", pa.float32()),\n",
    "    (\"Metadata_Protein\", pa.bool_()),\n",
    "    (\"Metadata_Control\", pa.bool_()),\n",
    "])\n",
    "writer = pq.ParquetWriter(preds_output_path, schema, compression=\"gzip\")\n",
    "\n",
    "try:\n",
    "    assert (\n",
    "        ~np.isnan(dframe[feat_col]).any().any()\n",
    "    ), \"Dataframe contains no NaN features.\"\n",
    "    assert (\n",
    "        np.isfinite(dframe[feat_col]).all().all()\n",
    "    ), \"Dataframe contains finite feature values.\"\n",
    "except AssertionError:\n",
    "    dframe = remove_nan_infs_columns(dframe)\n",
    "\n",
    "# Filter rows with NaN Metadata\n",
    "dframe = dframe[~dframe[\"Metadata_well_position\"].isna()]\n",
    "dframe = add_control_annot(dframe)\n",
    "dframe = dframe[~dframe[\"Metadata_control\"].isna()]\n",
    "\n",
    "# Split data into controls and alleles\n",
    "df_exp = dframe[~dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(drop=True)\n",
    "df_control = dframe[dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "# Remove any remaining TC from analysis\n",
    "df_control = df_control[df_control[\"Metadata_node_type\"] != \"TC\"].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "with open(logfile_path, \"w\") as log_file:\n",
    "    log_file.write(f\"===============================================================================================================================================================\\n\")\n",
    "    log_file.write(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "    print(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "\n",
    "    df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "    log_file.write(\"Dropping low cell count wells in ref. vs variant alleles:\\n\")\n",
    "    print(\"Dropping low cell count wells in ref. vs variant alleles:\")\n",
    "    df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "    log_file.write(f\"===============================================================================================================================================================\\n\\n\")\n",
    "\n",
    "    # Filter out wells with fewer than the cell count threhsold\n",
    "    df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "    df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "\n",
    "    # Protein feature analysis\n",
    "    df_feat_pro_con, df_result_pro_con = control_group_runner(\n",
    "        df_control, pq_writer=writer, log_file=log_file, protein=True\n",
    "    )\n",
    "    df_feat_pro_exp, df_result_pro_exp = experimental_runner(\n",
    "        df_exp, pq_writer=writer, log_file=log_file, protein=True\n",
    "    )\n",
    "\n",
    "    # Non-protein feature analysis\n",
    "    df_feat_no_pro_con, df_result_no_pro_con = control_group_runner(\n",
    "        df_control, pq_writer=writer, log_file=log_file, protein=False\n",
    "    )\n",
    "    df_feat_no_pro_exp, df_result_no_pro_exp = experimental_runner(\n",
    "        df_exp, pq_writer=writer, log_file=log_file, protein=False\n",
    "    )\n",
    "    writer.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varchamp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
