{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "419c9d9d",
   "metadata": {},
   "source": [
    "# Debugging the classify module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8fc2eab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Classification pipeline\"\"\"\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "from itertools import combinations\n",
    "from typing import Union\n",
    "import cupy as cp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "import xgboost as xgb\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from tqdm import tqdm\n",
    "from tqdm.contrib.concurrent import thread_map\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sys.path.append(\"..\")\n",
    "from utils import find_feat_cols, find_meta_cols, remove_nan_infs_columns\n",
    "\n",
    "\n",
    "def classifier(df_train, df_test, log_file, target=\"Label\", shuffle=False):\n",
    "    \"\"\"\n",
    "    This function runs classification.\n",
    "    \"\"\"\n",
    "\n",
    "    feat_col = find_feat_cols(df_train)\n",
    "    feat_col.remove(target)\n",
    "\n",
    "    x_train, y_train = cp.array(df_train[feat_col].to_numpy()), df_train[[target]]\n",
    "    x_test, y_test = cp.array(df_test[feat_col].to_numpy()), df_test[[target]]\n",
    "\n",
    "    num_pos = df_train[df_train[target] == 1].shape[0]\n",
    "    num_neg = df_train[df_train[target] == 0].shape[0]\n",
    "\n",
    "    if (num_pos == 0) or (num_neg == 0):\n",
    "        log_file.write(f\"Missing positive/negative labels in {df_train['Metadata_Plate'].unique()}, {df_train['Metadata_symbol'].unique()} wells: {df_train['Metadata_well_position'].unique()}\")\n",
    "        log_file.write(f\"Size of pos: {num_pos}, Size of neg: {num_neg}\\n\")\n",
    "\n",
    "        print(f\"size of pos: {num_pos}, size of neg: {num_neg}\")\n",
    "        feat_importances = pd.Series(np.nan, index=df_train[feat_col].columns)\n",
    "        return feat_importances, np.nan\n",
    "\n",
    "    scale_pos_weight = num_neg / num_pos\n",
    "\n",
    "    if (scale_pos_weight > 100) or (scale_pos_weight < 0.01):\n",
    "        log_file.write(f\"Extreme class imbalance in {df_train['Metadata_Plate'].unique()}, {df_train['Metadata_symbol'].unique()} wells: {df_train['Metadata_well_position'].unique()}\")\n",
    "        log_file.write(f\"Scale_pos_weight: {scale_pos_weight}, Size of pos: {num_pos}, Size of neg: {num_neg}\\n\")\n",
    "        print(\n",
    "            f\"scale_pos_weight: {scale_pos_weight}, size of pos: {num_pos}, size of neg: {num_neg}\"\n",
    "        )\n",
    "        feat_importances = pd.Series(np.nan, index=df_train[feat_col].columns)\n",
    "        return feat_importances, np.nan\n",
    "\n",
    "    le = LabelEncoder()\n",
    "    y_train = cp.array(le.fit_transform(y_train))\n",
    "    y_test = cp.array(le.fit_transform(y_test))\n",
    "\n",
    "    if shuffle:\n",
    "        # Create shuffled train labels\n",
    "        y_train_shuff = y_train.copy()\n",
    "        y_train_shuff[\"Label\"] = np.random.permutation(y_train.values)\n",
    "\n",
    "    model = xgb.XGBClassifier(\n",
    "        objective=\"binary:logistic\",\n",
    "        n_estimators=150,\n",
    "        tree_method=\"hist\",\n",
    "        device=\"cuda\",\n",
    "        learning_rate=0.05,\n",
    "        scale_pos_weight=scale_pos_weight,\n",
    "    ).fit(x_train, y_train, verbose=False)\n",
    "\n",
    "    # get predictions and scores\n",
    "    pred_score = model.predict_proba(x_test)[:, 1]\n",
    "\n",
    "    # Return classifier info\n",
    "    info_0 = df_test[df_test[\"Label\"] == 0].iloc[0]\n",
    "    info_1 = df_test[df_test[\"Label\"] == 1].iloc[0]\n",
    "    class_ID = (\n",
    "        info_0[\"Metadata_Plate\"]\n",
    "        + \"_\"\n",
    "        + info_0[\"Metadata_well_position\"]\n",
    "        + \"_\"\n",
    "        + info_1[\"Metadata_well_position\"]\n",
    "    )\n",
    "    classifier_df = pd.DataFrame({\n",
    "        \"Classifier_ID\": [class_ID],\n",
    "        \"Plate\": [info_0[\"Metadata_Plate\"]],\n",
    "        \"trainsize_0\": [sum(y_train.get() == 0)],\n",
    "        \"testsize_0\": [sum(y_test.get() == 0)],\n",
    "        \"well_0\": [info_0[\"Metadata_well_position\"]],\n",
    "        \"allele_0\": [info_0[\"Metadata_gene_allele\"]],\n",
    "        \"trainsize_1\": [sum(y_train.get() == 1)],\n",
    "        \"testsize_1\": [sum(y_test.get() == 1)],\n",
    "        \"well_1\": [info_1[\"Metadata_well_position\"]],\n",
    "        \"allele_1\": [info_1[\"Metadata_gene_allele\"]],\n",
    "    })\n",
    "\n",
    "    # Store feature importance\n",
    "    feat_importances = pd.Series(\n",
    "        model.feature_importances_, index=df_train[feat_col].columns\n",
    "    )\n",
    "\n",
    "    # Return cell-level predictions\n",
    "    cellID = df_test.apply(\n",
    "        lambda row: f\"{row['Metadata_Plate']}_{row['Metadata_well_position']}_{row['Metadata_ImageNumber']}_{row['Metadata_ObjectNumber']}\",\n",
    "        axis=1,\n",
    "    ).to_list()\n",
    "\n",
    "    pred_df = pd.DataFrame({\n",
    "        \"Classifier_ID\": class_ID,\n",
    "        \"CellID\": cellID,\n",
    "        \"Label\": y_test.get(),\n",
    "        \"Prediction\": pred_score,\n",
    "    })\n",
    "\n",
    "    return feat_importances, classifier_df, pred_df\n",
    "\n",
    "\n",
    "def get_classifier_features(dframe: pd.DataFrame, protein_feat: bool):\n",
    "    \"\"\"Helper function to get dframe containing protein or non-protein features\"\"\"\n",
    "    feat_col = find_feat_cols(dframe)\n",
    "    meta_col = find_meta_cols(dframe)\n",
    "\n",
    "    if protein_feat:\n",
    "        feat_col = [\n",
    "            i\n",
    "            for i in feat_col\n",
    "            if (\"GFP\" in i)\n",
    "            and (\"DNA\" not in i)\n",
    "            and (\"AGP\" not in i)\n",
    "            and (\"Mito\" not in i)\n",
    "            and (\"Brightfield\" not in i)\n",
    "        ]\n",
    "    else:\n",
    "        feat_col = [\n",
    "            i\n",
    "            for i in feat_col\n",
    "            if (\"GFP\" not in i) and (\"Brightfield\" not in i) ## and (\"AGP\" not in i), test without AGP features\n",
    "        ]\n",
    "\n",
    "    dframe = pd.concat([dframe[meta_col], dframe[feat_col]], axis=1)\n",
    "    return dframe\n",
    "\n",
    "\n",
    "def stratify_by_plate(df_sampled: pd.DataFrame, plate: str):\n",
    "    \"\"\"Stratify dframe by plate\"\"\"\n",
    "    # print(df_sampled.head())\n",
    "    df_sampled_platemap = plate.split(\"_T\")[0]\n",
    "    platemaps = df_sampled[df_sampled[\"Metadata_Plate\"].str.contains(df_sampled_platemap)][\"Metadata_plate_map_name\"].to_list()\n",
    "    assert(len(set(platemaps))==1), \"Only one platemap should be associated with plate: {plate}.\"\n",
    "    platemap = platemaps[0]\n",
    "\n",
    "    # Train on data from same platemap but other plates\n",
    "    df_train = df_sampled[\n",
    "        (df_sampled[\"Metadata_plate_map_name\"] == platemap)\n",
    "        & (df_sampled[\"Metadata_Plate\"] != plate)\n",
    "    ].reset_index(drop=True)\n",
    "\n",
    "    df_test = df_sampled[df_sampled[\"Metadata_Plate\"] == plate].reset_index(drop=True)\n",
    "    # print(df_train.head())\n",
    "    # print(df_test.head())\n",
    "\n",
    "    return df_train, df_test\n",
    "\n",
    "\n",
    "def experimental_runner(\n",
    "    exp_dframe: pd.DataFrame,\n",
    "    pq_writer,\n",
    "    log_file,\n",
    "    protein=True,\n",
    "    group_key_one=\"Metadata_symbol\",\n",
    "    group_key_two=\"Metadata_gene_allele\",\n",
    "    threshold_key=\"Metadata_node_type\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Run Reference v.s. Variant experiments\n",
    "    \"\"\"\n",
    "    exp_dframe = get_classifier_features(exp_dframe, protein)\n",
    "    feat_cols = find_feat_cols(exp_dframe)\n",
    "    feat_cols = [i for i in feat_cols if i != \"Label\"]\n",
    "\n",
    "    group_list = []\n",
    "    pair_list = []\n",
    "    feat_list = []\n",
    "    info_list = []\n",
    "\n",
    "    log_file.write(f\"Running XGBboost classifiers w/ protein {protein} on target variants:\\n\")\n",
    "    groups = exp_dframe.groupby(group_key_one).groups\n",
    "    for key in tqdm(groups.keys()):\n",
    "        dframe_grouped = exp_dframe.loc[groups[key]].reset_index(drop=True)\n",
    "\n",
    "        # Ensure this gene has both reference and variants\n",
    "        if dframe_grouped[threshold_key].unique().size < 2:\n",
    "            continue\n",
    "\n",
    "        df_group_one = dframe_grouped[\n",
    "            dframe_grouped[threshold_key] == \"disease_wt\"\n",
    "        ].reset_index(drop=True)\n",
    "        df_group_one[\"Label\"] = 1\n",
    "\n",
    "        subgroups = (\n",
    "            dframe_grouped[dframe_grouped[threshold_key] == \"allele\"]\n",
    "            .groupby(group_key_two)\n",
    "            .groups\n",
    "        )\n",
    "\n",
    "        for subkey in subgroups.keys():\n",
    "            df_group_two = dframe_grouped.loc[subgroups[subkey]].reset_index(drop=True)\n",
    "            df_group_two[\"Label\"] = 0\n",
    "            plate_list = get_common_plates(df_group_one, df_group_two)\n",
    "\n",
    "            ref_wells = df_group_one[\"Metadata_well_position\"].unique()\n",
    "            var_wells = list(df_group_two[\"Metadata_well_position\"].unique())\n",
    "            ref_var_pairs = [(ref_well, var_well) for ref_well in ref_wells for var_well in var_wells]\n",
    "            df_sampled_ = pd.concat([df_group_one, df_group_two], ignore_index=True)\n",
    "\n",
    "            for ref_var in ref_var_pairs:\n",
    "                df_sampled = df_sampled_[df_sampled_[\"Metadata_well_position\"].isin(ref_var)]\n",
    "                def classify_by_plate_helper(plate):\n",
    "                    df_train, df_test = stratify_by_plate(df_sampled, plate)\n",
    "                    feat_importances, classifier_info, predictions = classifier(\n",
    "                        df_train, df_test, log_file\n",
    "                    )\n",
    "                    return {plate: [feat_importances, classifier_info, predictions]}\n",
    "\n",
    "                # try run classifier\n",
    "                try:\n",
    "                    result = thread_map(classify_by_plate_helper, plate_list)\n",
    "                    pred_list = []\n",
    "                    for res in result:\n",
    "                        if len(list(res.values())[0]) == 3:\n",
    "                            feat_list.append(list(res.values())[0][0])\n",
    "                            group_list.append(key)\n",
    "                            pair_list.append(f\"{key}_{subkey}\")\n",
    "                            info_list.append(list(res.values())[0][1])\n",
    "                            pred_list.append(list(res.values())[0][2])\n",
    "                        else:\n",
    "                            print(\"res length not 3!\")\n",
    "                            feat_list.append([None] * len(feat_cols))\n",
    "                            group_list.append(key)\n",
    "                            pair_list.append(f\"{key}_{subkey}\")\n",
    "                            info_list.append([None] * 10)\n",
    "\n",
    "                    cell_preds = pd.concat(pred_list, axis=0)\n",
    "                    cell_preds[\"Metadata_Protein\"] = protein\n",
    "                    cell_preds[\"Metadata_Control\"] = False\n",
    "                    table = pa.Table.from_pandas(cell_preds, preserve_index=False)\n",
    "                    pq_writer.write_table(table)\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "                    log_file.write(f\"{key}, {subkey} error: {e}\\n\")\n",
    "            break\n",
    "        break\n",
    "\n",
    "    # Store feature importance\n",
    "    df_feat_one = pd.DataFrame({\"Group1\": group_list, \"Group2\": pair_list})\n",
    "    df_feat_two = pd.DataFrame(feat_list)\n",
    "    df_feat = pd.concat([df_feat_one, df_feat_two], axis=1)\n",
    "    df_feat[\"Metadata_Protein\"] = protein\n",
    "    df_feat[\"Metadata_Control\"] = False\n",
    "\n",
    "    # process classifier info\n",
    "    df_result = pd.concat(info_list, ignore_index=True)\n",
    "    df_result[\"Metadata_Control\"] = False\n",
    "\n",
    "    log_file.write(f\"Finished running XGBboost classifiers w/ protein {protein} on target variants.\\n\")\n",
    "    log_file.write(f\"===========================================================================\\n\\n\")\n",
    "    return df_feat, df_result\n",
    "\n",
    "\n",
    "def get_common_plates(dframe1, dframe2):\n",
    "    \"\"\"Helper func: get common plates in two dataframes\"\"\"\n",
    "    plate_list = list(\n",
    "        set(list(dframe1[\"Metadata_Plate\"].unique()))\n",
    "        & set(list(dframe2[\"Metadata_Plate\"].unique()))\n",
    "    )\n",
    "    return plate_list\n",
    "\n",
    "\n",
    "def control_group_runner(\n",
    "    ctrl_dframe: pd.DataFrame,\n",
    "    pq_writer,\n",
    "    log_file,\n",
    "    group_key_one=\"Metadata_gene_allele\",\n",
    "    group_key_two=\"Metadata_plate_map_name\",\n",
    "    group_key_three=\"Metadata_well_position\",\n",
    "    threshold_key=\"Metadata_well_position\",\n",
    "    protein=True,\n",
    "):\n",
    "    \"\"\"\n",
    "    Run null control experiments.\n",
    "    \"\"\"\n",
    "    ctrl_dframe = get_classifier_features(ctrl_dframe, protein)\n",
    "    feat_cols = find_feat_cols(ctrl_dframe)\n",
    "    feat_cols = [i for i in feat_cols if i != \"Label\"]\n",
    "\n",
    "    group_list = []\n",
    "    pair_list = []\n",
    "    feat_list = []\n",
    "    info_list = []\n",
    "\n",
    "    log_file.write(f\"Running XGBboost classifiers w/ protein {protein} on control alleles:\\n\")\n",
    "    groups = ctrl_dframe.groupby(group_key_one).groups\n",
    "    for key in tqdm(groups.keys()):\n",
    "        # groupby alleles\n",
    "        dframe_grouped = ctrl_dframe.loc[groups[key]].reset_index(drop=True)\n",
    "\n",
    "        # Skip controls with no replicates\n",
    "        if dframe_grouped[threshold_key].unique().size < 2:\n",
    "            continue\n",
    "\n",
    "        # group by platemap\n",
    "        subgroups = dframe_grouped.groupby(group_key_two).groups\n",
    "        for key_two in subgroups.keys():\n",
    "            dframe_grouped_two = dframe_grouped.loc[subgroups[key_two]].reset_index(\n",
    "                drop=True\n",
    "            )\n",
    "            # If a well is not present on all four plates, drop well\n",
    "            well_count = dframe_grouped_two.groupby([\"Metadata_Well\"])[\n",
    "                \"Metadata_Plate\"\n",
    "            ].nunique()\n",
    "            well_to_drop = well_count[well_count < 4].index\n",
    "            dframe_grouped_two = dframe_grouped_two[\n",
    "                ~dframe_grouped_two[\"Metadata_Well\"].isin(well_to_drop)\n",
    "            ].reset_index(drop=True)\n",
    "\n",
    "            # group by well\n",
    "            sub_sub_groups = dframe_grouped_two.groupby(group_key_three).groups\n",
    "            sampled_pairs = list(combinations(list(sub_sub_groups.keys()), r=2))\n",
    "\n",
    "            for idx1, idx2 in sampled_pairs:\n",
    "                df_group_one = dframe_grouped_two.loc[sub_sub_groups[idx1]].reset_index(\n",
    "                    drop=True\n",
    "                )\n",
    "                df_group_one[\"Label\"] = 1\n",
    "                df_group_two = dframe_grouped_two.loc[sub_sub_groups[idx2]].reset_index(\n",
    "                    drop=True\n",
    "                )\n",
    "                df_group_two[\"Label\"] = 0\n",
    "                df_sampled = pd.concat([df_group_one, df_group_two], ignore_index=True)\n",
    "\n",
    "                try:\n",
    "                    plate_list = get_common_plates(df_group_one, df_group_two)\n",
    "                    def classify_by_plate_helper(plate):\n",
    "                        df_train, df_test = stratify_by_plate(df_sampled, plate)\n",
    "                        feat_importances, classifier_info, predictions = classifier(\n",
    "                            df_train, df_test, log_file\n",
    "                        )\n",
    "                        return {plate: [feat_importances, classifier_info, predictions]}\n",
    "\n",
    "                    result = thread_map(classify_by_plate_helper, plate_list)\n",
    "\n",
    "                    pred_list = []\n",
    "                    for res in result:\n",
    "                        if len(list(res.values())[0]) == 3:\n",
    "                            feat_list.append(list(res.values())[0][0])\n",
    "                            group_list.append(key)\n",
    "                            pair_list.append(f\"{idx1}_{idx2}\")\n",
    "                            info_list.append(list(res.values())[0][1])\n",
    "                            pred_list.append(list(res.values())[0][2])\n",
    "                        else:\n",
    "                            print(\"res length does not equal three!\")\n",
    "                            feat_list.append([None] * len(feat_cols))\n",
    "                            group_list.append(key)\n",
    "                            pair_list.append(f\"{idx1}_{idx2}\")\n",
    "                            info_list.append([None] * 10)\n",
    "\n",
    "                    cell_preds = pd.concat(pred_list, axis=0)\n",
    "                    cell_preds[\"Metadata_Protein\"] = protein\n",
    "                    cell_preds[\"Metadata_Control\"] = True\n",
    "                    table = pa.Table.from_pandas(cell_preds, preserve_index=False)\n",
    "                    pq_writer.write_table(table)\n",
    "                except Exception as e:\n",
    "                    print(e)\n",
    "                    log_file.write(f\"{key}, {key_two} error: {e}, wells per ctrl: {sub_sub_groups}\\n\")\n",
    "                break\n",
    "            break\n",
    "\n",
    "    # Store feature importance\n",
    "    df_feat_one = pd.DataFrame({\"Group1\": group_list, \"Group2\": pair_list})\n",
    "    df_feat_two = pd.DataFrame(feat_list)\n",
    "    df_feat = pd.concat([df_feat_one, df_feat_two], axis=1)\n",
    "    df_feat[\"Metadata_Protein\"] = protein\n",
    "    df_feat[\"Metadata_Control\"] = True\n",
    "\n",
    "    # process classifier info\n",
    "    df_result = pd.concat(info_list, ignore_index=True)\n",
    "    df_result[\"Metadata_Control\"] = True\n",
    "\n",
    "    log_file.write(f\"Finished running XGBboost classifiers w/ protein {protein} on control alleles.\\n\")\n",
    "    log_file.write(f\"===========================================================================\\n\\n\")\n",
    "    return df_feat, df_result\n",
    "\n",
    "\n",
    "def control_type_helper(col_annot: str):\n",
    "    \"\"\"helper func for annotating column \"Metadata_control\" \"\"\"\n",
    "    ## Only TC, NC, PC are used for constructing the null distribution because of multiple duplicates \n",
    "    if col_annot in [\"TC\", \"NC\", \"PC\"]:\n",
    "        return True\n",
    "    ## else labeled as not controls\n",
    "    elif col_annot in [\"disease_wt\", \"allele\", \"cPC\", \"cNC\"]:\n",
    "        return False\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def add_control_annot(dframe):\n",
    "    \"\"\"annotating column \"Metadata_control\" \"\"\"\n",
    "    if \"Metadata_control\" not in dframe.columns:\n",
    "        dframe[\"Metadata_control\"] = dframe[\"Metadata_node_type\"].apply(\n",
    "            lambda x: control_type_helper(x)\n",
    "        )\n",
    "    return dframe\n",
    "\n",
    "\n",
    "def drop_low_cc_wells(dframe, cc_thresh, log_file):\n",
    "    # Drop wells with cell counts lower than the threshold\n",
    "    dframe[\"Metadata_Cell_ID\"] = dframe.index\n",
    "    cell_count = (\n",
    "        dframe.groupby([\"Metadata_Plate\", \"Metadata_Well\"])[\"Metadata_Cell_ID\"]\n",
    "        .count()\n",
    "        .reset_index(name=\"Metadata_Cell_Count\")\n",
    "    )\n",
    "    ## get the cell counts per well per plate\n",
    "    dframe = dframe.merge(\n",
    "        cell_count,\n",
    "        on=[\"Metadata_Plate\", \"Metadata_Well\"],\n",
    "    )\n",
    "    dframe_dropped = (\n",
    "        dframe[dframe[\"Metadata_Cell_Count\"] < cc_thresh]\n",
    "    )\n",
    "    ## keep track of the alleles in a log file\n",
    "    log_file.write(f\"Number of wells dropped due to cell counts < {cc_thresh}: {len((dframe_dropped['Metadata_Plate']+dframe_dropped['Metadata_Well']+dframe_dropped['Metadata_gene_allele']).unique())}\\n\")\n",
    "    dframe_dropped = dframe_dropped.drop_duplicates(subset=[\"Metadata_Plate\", \"Metadata_Well\"])\n",
    "    if (dframe_dropped.shape[0] > 0):\n",
    "        for idx in dframe_dropped.index:\n",
    "            log_file.write(f\"{dframe_dropped.loc[idx, 'Metadata_Plate']}, {dframe_dropped.loc[idx, 'Metadata_Well']}:{dframe_dropped.loc[idx, 'Metadata_gene_allele']}\\n\")\n",
    "            # print(f\"{dframe_dropped.loc[idx, 'Metadata_Plate']}, {dframe_dropped.loc[idx, 'Metadata_Well']}:{dframe_dropped.loc[idx, 'Metadata_gene_allele']}\\n\")\n",
    "    ## keep only the wells with cc >= cc_thresh\n",
    "    dframe = (\n",
    "        dframe[dframe[\"Metadata_Cell_Count\"] >= cc_thresh]\n",
    "        .drop(columns=[\"Metadata_Cell_Count\"])\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "    return dframe\n",
    "\n",
    "\n",
    "def run_classify_workflow(\n",
    "    input_path: str,\n",
    "    feat_output_path: str,\n",
    "    info_output_path: str,\n",
    "    preds_output_path: str,\n",
    "    cc_threshold: int,\n",
    "    use_gpu: Union[str, None] = \"0,1\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Run workflow for single-cell classification\n",
    "    \"\"\"\n",
    "    if use_gpu is not None:\n",
    "        os.environ[\"CUDA_VISIBLE_DEVICES\"] = use_gpu\n",
    "\n",
    "    # Initialize parquet for cell-level predictions\n",
    "    if os.path.exists(preds_output_path):\n",
    "        os.remove(preds_output_path)\n",
    "\n",
    "    logfile_path = os.path.join(*[preds_output_path.split(\"/\")[:-1], \"classify.log\"])\n",
    "\n",
    "    schema = pa.schema([\n",
    "        (\"Classifier_ID\", pa.string()),\n",
    "        (\"CellID\", pa.string()),\n",
    "        (\"Label\", pa.int64()),\n",
    "        (\"Prediction\", pa.float32()),\n",
    "        (\"Metadata_Protein\", pa.bool_()),\n",
    "        (\"Metadata_Control\", pa.bool_()),\n",
    "    ])\n",
    "    writer = pq.ParquetWriter(preds_output_path, schema, compression=\"gzip\")\n",
    "\n",
    "    # Add CellID column\n",
    "    dframe = (\n",
    "        pl.scan_parquet(input_path)\n",
    "        .with_columns(\n",
    "            pl.concat_str(\n",
    "                [\n",
    "                    \"Metadata_Plate\",\n",
    "                    \"Metadata_well_position\",\n",
    "                    \"Metadata_ImageNumber\",\n",
    "                    \"Metadata_ObjectNumber\",\n",
    "                ],\n",
    "                separator=\"_\",\n",
    "            ).alias(\"Metadata_CellID\")\n",
    "        )\n",
    "        .collect()\n",
    "        .to_pandas()\n",
    "    )\n",
    "\n",
    "    feat_col = find_feat_cols(dframe)\n",
    "\n",
    "    try:\n",
    "        assert (\n",
    "            ~np.isnan(dframe[feat_col]).any().any()\n",
    "        ), \"Dataframe contains no NaN features.\"\n",
    "        assert (\n",
    "            np.isfinite(dframe[feat_col]).all().all()\n",
    "        ), \"Dataframe contains finite feature values.\"\n",
    "    except AssertionError:\n",
    "        dframe = remove_nan_infs_columns(dframe)\n",
    "\n",
    "    # Filter rows with NaN Metadata\n",
    "    dframe = dframe[~dframe[\"Metadata_well_position\"].isna()]\n",
    "    dframe = add_control_annot(dframe)\n",
    "    dframe = dframe[~dframe[\"Metadata_control\"].isna()]\n",
    "\n",
    "    # Split data into controls and alleles\n",
    "    df_exp = dframe[~dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(drop=True)\n",
    "    df_control = dframe[dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(\n",
    "        drop=True\n",
    "    )\n",
    "\n",
    "    # Remove any remaining TC from analysis\n",
    "    df_control = df_control[df_control[\"Metadata_node_type\"] != \"TC\"].reset_index(\n",
    "        drop=True\n",
    "    )\n",
    "\n",
    "    with open(logfile_path, \"w\") as log_file:\n",
    "        log_file.write(f\"===============================================================================================================================================================\\n\")\n",
    "        log_file.write(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "        print(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "\n",
    "        df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "        log_file.write(\"Dropping low cell count wells in ref. vs variant alleles:\\n\")\n",
    "        print(\"Dropping low cell count wells in ref. vs variant alleles:\")\n",
    "        df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "        log_file.write(f\"===============================================================================================================================================================\\n\\n\")\n",
    "\n",
    "        # Filter out wells with fewer than the cell count threhsold\n",
    "        df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "        df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "\n",
    "        # Protein feature analysis\n",
    "        df_feat_pro_con, df_result_pro_con = control_group_runner(\n",
    "            df_control, pq_writer=writer, log_file=log_file, protein=True\n",
    "        )\n",
    "        df_feat_pro_exp, df_result_pro_exp = experimental_runner(\n",
    "            df_exp, pq_writer=writer, log_file=log_file, protein=True\n",
    "        )\n",
    "\n",
    "        # Non-protein feature analysis\n",
    "        df_feat_no_pro_con, df_result_no_pro_con = control_group_runner(\n",
    "            df_control, pq_writer=writer, log_file=log_file, protein=False\n",
    "        )\n",
    "        df_feat_no_pro_exp, df_result_no_pro_exp = experimental_runner(\n",
    "            df_exp, pq_writer=writer, log_file=log_file, protein=False\n",
    "        )\n",
    "        writer.close()\n",
    "\n",
    "    # Concatenate results for both protein and non-protein\n",
    "    df_feat = pd.concat(\n",
    "        [df_feat_pro_con, df_feat_no_pro_con, df_feat_pro_exp, df_feat_no_pro_exp],\n",
    "        ignore_index=True,\n",
    "    )\n",
    "    df_result = pd.concat(\n",
    "        [\n",
    "            df_result_pro_con,\n",
    "            df_result_no_pro_con,\n",
    "            df_result_pro_exp,\n",
    "            df_result_no_pro_exp,\n",
    "        ],\n",
    "        ignore_index=True,\n",
    "    )\n",
    "    df_result = df_result.drop_duplicates()\n",
    "\n",
    "    # Write out feature importance and classifier info\n",
    "    df_feat.to_csv(feat_output_path, index=False)\n",
    "    df_result.to_csv(info_output_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f245a91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = \"../outputs/batch_profiles/2025_01_27_Batch_13/profiles_tcdropped_filtered_var_mad_outlier_featselect.parquet\"\n",
    "# Add CellID column\n",
    "dframe = (\n",
    "    pl.scan_parquet(input_path)\n",
    "    .with_columns(\n",
    "        pl.concat_str(\n",
    "            [\n",
    "                \"Metadata_Plate\",\n",
    "                \"Metadata_well_position\",\n",
    "                \"Metadata_ImageNumber\",\n",
    "                \"Metadata_ObjectNumber\",\n",
    "            ],\n",
    "            separator=\"_\",\n",
    "        ).alias(\"Metadata_CellID\")\n",
    "    )\n",
    "    .collect()\n",
    "    .filter(pl.col(\"Metadata_symbol\")==\"CCM2\")\n",
    "    .to_pandas()\n",
    ")\n",
    "feat_col = find_feat_cols(dframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3134f0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_plate_map_name</th>\n",
       "      <th>Metadata_well_position</th>\n",
       "      <th>Metadata_symbol</th>\n",
       "      <th>Metadata_gene_allele</th>\n",
       "      <th>Metadata_imaging_well</th>\n",
       "      <th>Metadata_imaging_plate_R1</th>\n",
       "      <th>Metadata_imaging_plate_R2</th>\n",
       "      <th>Metadata_node_type</th>\n",
       "      <th>Metadata_orf_id_wt</th>\n",
       "      <th>Metadata_ccsb_mutation_id</th>\n",
       "      <th>...</th>\n",
       "      <th>Cells_RadialDistribution_FracAtD_mito_tubeness_7of20</th>\n",
       "      <th>Nuclei_ObjectSkeleton_TotalObjectSkeletonLength_mito_skel</th>\n",
       "      <th>Cytoplasm_RadialDistribution_RadialCV_DNA_4of10</th>\n",
       "      <th>Cells_RadialDistribution_RadialCV_GFP_3of10</th>\n",
       "      <th>Nuclei_RadialDistribution_FracAtD_DNA_6of10</th>\n",
       "      <th>Nuclei_Texture_Contrast_AGP_20_00_256</th>\n",
       "      <th>Nuclei_Texture_SumAverage_GFP_5_01_256</th>\n",
       "      <th>Cytoplasm_Texture_InfoMeas1_Mito_20_02_256</th>\n",
       "      <th>Cytoplasm_Granularity_1_AGP</th>\n",
       "      <th>Metadata_CellID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>M10</td>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>B14A7A8P2_R2</td>\n",
       "      <td>disease_wt</td>\n",
       "      <td>3928.0</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>2.239665</td>\n",
       "      <td>0.799854</td>\n",
       "      <td>3.747274</td>\n",
       "      <td>-0.160807</td>\n",
       "      <td>-2.421128</td>\n",
       "      <td>0.825395</td>\n",
       "      <td>-0.262621</td>\n",
       "      <td>-2.584699</td>\n",
       "      <td>-0.302897</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_M10_2676_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>M10</td>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>B14A7A8P2_R2</td>\n",
       "      <td>disease_wt</td>\n",
       "      <td>3928.0</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>1.469021</td>\n",
       "      <td>-0.796329</td>\n",
       "      <td>1.919917</td>\n",
       "      <td>-0.198846</td>\n",
       "      <td>0.001183</td>\n",
       "      <td>-0.498640</td>\n",
       "      <td>-1.013328</td>\n",
       "      <td>-0.398445</td>\n",
       "      <td>-2.208567</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_M10_2677_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>M10</td>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>B14A7A8P2_R2</td>\n",
       "      <td>disease_wt</td>\n",
       "      <td>3928.0</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.130079</td>\n",
       "      <td>0.175046</td>\n",
       "      <td>1.994589</td>\n",
       "      <td>-0.937285</td>\n",
       "      <td>0.343746</td>\n",
       "      <td>0.124780</td>\n",
       "      <td>-0.433273</td>\n",
       "      <td>-1.471504</td>\n",
       "      <td>0.066759</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_M10_2678_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>M10</td>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>B14A7A8P2_R2</td>\n",
       "      <td>disease_wt</td>\n",
       "      <td>3928.0</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>1.094945</td>\n",
       "      <td>2.568591</td>\n",
       "      <td>-2.042540</td>\n",
       "      <td>0.553863</td>\n",
       "      <td>-0.399926</td>\n",
       "      <td>0.771891</td>\n",
       "      <td>0.125791</td>\n",
       "      <td>1.048846</td>\n",
       "      <td>10.353196</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_M10_2679_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>M10</td>\n",
       "      <td>B13A7A8P2_R1</td>\n",
       "      <td>B14A7A8P2_R2</td>\n",
       "      <td>disease_wt</td>\n",
       "      <td>3928.0</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.429298</td>\n",
       "      <td>0.728380</td>\n",
       "      <td>4.118041</td>\n",
       "      <td>-0.258514</td>\n",
       "      <td>1.336622</td>\n",
       "      <td>2.344744</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>1.436391</td>\n",
       "      <td>1.827263</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_M10_2680_6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1163 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_plate_map_name Metadata_well_position Metadata_symbol  \\\n",
       "0            B13A7A8P2_R1                    M10            CCM2   \n",
       "1            B13A7A8P2_R1                    M10            CCM2   \n",
       "2            B13A7A8P2_R1                    M10            CCM2   \n",
       "3            B13A7A8P2_R1                    M10            CCM2   \n",
       "4            B13A7A8P2_R1                    M10            CCM2   \n",
       "\n",
       "  Metadata_gene_allele Metadata_imaging_well Metadata_imaging_plate_R1  \\\n",
       "0                 CCM2                   M10              B13A7A8P2_R1   \n",
       "1                 CCM2                   M10              B13A7A8P2_R1   \n",
       "2                 CCM2                   M10              B13A7A8P2_R1   \n",
       "3                 CCM2                   M10              B13A7A8P2_R1   \n",
       "4                 CCM2                   M10              B13A7A8P2_R1   \n",
       "\n",
       "  Metadata_imaging_plate_R2 Metadata_node_type  Metadata_orf_id_wt  \\\n",
       "0              B14A7A8P2_R2         disease_wt              3928.0   \n",
       "1              B14A7A8P2_R2         disease_wt              3928.0   \n",
       "2              B14A7A8P2_R2         disease_wt              3928.0   \n",
       "3              B14A7A8P2_R2         disease_wt              3928.0   \n",
       "4              B14A7A8P2_R2         disease_wt              3928.0   \n",
       "\n",
       "  Metadata_ccsb_mutation_id  ...  \\\n",
       "0                      None  ...   \n",
       "1                      None  ...   \n",
       "2                      None  ...   \n",
       "3                      None  ...   \n",
       "4                      None  ...   \n",
       "\n",
       "  Cells_RadialDistribution_FracAtD_mito_tubeness_7of20  \\\n",
       "0                                           2.239665     \n",
       "1                                           1.469021     \n",
       "2                                          -1.130079     \n",
       "3                                           1.094945     \n",
       "4                                          -1.429298     \n",
       "\n",
       "  Nuclei_ObjectSkeleton_TotalObjectSkeletonLength_mito_skel  \\\n",
       "0                                           0.799854          \n",
       "1                                          -0.796329          \n",
       "2                                           0.175046          \n",
       "3                                           2.568591          \n",
       "4                                           0.728380          \n",
       "\n",
       "   Cytoplasm_RadialDistribution_RadialCV_DNA_4of10  \\\n",
       "0                                         3.747274   \n",
       "1                                         1.919917   \n",
       "2                                         1.994589   \n",
       "3                                        -2.042540   \n",
       "4                                         4.118041   \n",
       "\n",
       "   Cells_RadialDistribution_RadialCV_GFP_3of10  \\\n",
       "0                                    -0.160807   \n",
       "1                                    -0.198846   \n",
       "2                                    -0.937285   \n",
       "3                                     0.553863   \n",
       "4                                    -0.258514   \n",
       "\n",
       "   Nuclei_RadialDistribution_FracAtD_DNA_6of10  \\\n",
       "0                                    -2.421128   \n",
       "1                                     0.001183   \n",
       "2                                     0.343746   \n",
       "3                                    -0.399926   \n",
       "4                                     1.336622   \n",
       "\n",
       "   Nuclei_Texture_Contrast_AGP_20_00_256  \\\n",
       "0                               0.825395   \n",
       "1                              -0.498640   \n",
       "2                               0.124780   \n",
       "3                               0.771891   \n",
       "4                               2.344744   \n",
       "\n",
       "   Nuclei_Texture_SumAverage_GFP_5_01_256  \\\n",
       "0                               -0.262621   \n",
       "1                               -1.013328   \n",
       "2                               -0.433273   \n",
       "3                                0.125791   \n",
       "4                                0.000098   \n",
       "\n",
       "   Cytoplasm_Texture_InfoMeas1_Mito_20_02_256  Cytoplasm_Granularity_1_AGP  \\\n",
       "0                                   -2.584699                    -0.302897   \n",
       "1                                   -0.398445                    -2.208567   \n",
       "2                                   -1.471504                     0.066759   \n",
       "3                                    1.048846                    10.353196   \n",
       "4                                    1.436391                     1.827263   \n",
       "\n",
       "                      Metadata_CellID  \n",
       "0  2025_01_27_B13A7A8P2_T1_M10_2676_6  \n",
       "1  2025_01_27_B13A7A8P2_T1_M10_2677_6  \n",
       "2  2025_01_27_B13A7A8P2_T1_M10_2678_6  \n",
       "3  2025_01_27_B13A7A8P2_T1_M10_2679_6  \n",
       "4  2025_01_27_B13A7A8P2_T1_M10_2680_6  \n",
       "\n",
       "[5 rows x 1163 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(dframe.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8a6ec66f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping low cell count wells in control alleles:\n",
      "\n",
      "Dropping low cell count wells in ref. vs variant alleles:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f467477b088f4e1f81143dc5bbb9f537",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "665a15d5a2ea4ffb9b9ea5aa0625e9dd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4b1dfe3a72e74759b5fb2b5b0ddbdc15",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1 [00:07<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "# Initialize parquet for cell-level predictions\n",
    "batch = \"2025_01_27_Batch_13\"\n",
    "pipeline = \"profiles_tcdropped_filtered_var_mad_outlier_featselect_filtcells\"\n",
    "preds_output_path = f\"../outputs/results/{batch}/{pipeline}/predictions.parquet\"\n",
    "cc_threshold = 20\n",
    "\n",
    "if os.path.exists(preds_output_path):\n",
    "    os.remove(preds_output_path)\n",
    "\n",
    "if not os.path.exists(f\"../outputs/results/{batch}/{pipeline}\"):\n",
    "    os.makedirs(f\"../outputs/results/{batch}/{pipeline}\")\n",
    "\n",
    "logfile_path = os.path.join('/'.join(preds_output_path.split(\"/\")[:-1]), \"classify.log\")\n",
    "# logfile_path\n",
    "\n",
    "schema = pa.schema([\n",
    "    (\"Classifier_ID\", pa.string()),\n",
    "    (\"CellID\", pa.string()),\n",
    "    (\"Label\", pa.int64()),\n",
    "    (\"Prediction\", pa.float32()),\n",
    "    (\"Metadata_Protein\", pa.bool_()),\n",
    "    (\"Metadata_Control\", pa.bool_()),\n",
    "])\n",
    "writer = pq.ParquetWriter(preds_output_path, schema, compression=\"gzip\")\n",
    "\n",
    "try:\n",
    "    assert (\n",
    "        ~np.isnan(dframe[feat_col]).any().any()\n",
    "    ), \"Dataframe contains no NaN features.\"\n",
    "    assert (\n",
    "        np.isfinite(dframe[feat_col]).all().all()\n",
    "    ), \"Dataframe contains finite feature values.\"\n",
    "except AssertionError:\n",
    "    dframe = remove_nan_infs_columns(dframe)\n",
    "\n",
    "# Filter rows with NaN Metadata\n",
    "dframe = dframe[~dframe[\"Metadata_well_position\"].isna()]\n",
    "dframe = add_control_annot(dframe)\n",
    "dframe = dframe[~dframe[\"Metadata_control\"].isna()]\n",
    "\n",
    "# Split data into controls and alleles\n",
    "df_exp = dframe[~dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(drop=True)\n",
    "df_control = dframe[dframe[\"Metadata_control\"].astype(\"bool\")].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "# Remove any remaining TC from analysis\n",
    "df_control = df_control[df_control[\"Metadata_node_type\"] != \"TC\"].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "with open(logfile_path, \"w\") as log_file:\n",
    "    log_file.write(f\"===============================================================================================================================================================\\n\")\n",
    "    log_file.write(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "    print(\"Dropping low cell count wells in control alleles:\\n\")\n",
    "\n",
    "    df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "    log_file.write(\"Dropping low cell count wells in ref. vs variant alleles:\\n\")\n",
    "    print(\"Dropping low cell count wells in ref. vs variant alleles:\")\n",
    "    df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "    log_file.write(f\"===============================================================================================================================================================\\n\\n\")\n",
    "\n",
    "    # Filter out wells with fewer than the cell count threhsold\n",
    "    df_control = drop_low_cc_wells(df_control, cc_threshold, log_file)\n",
    "    df_exp = drop_low_cc_wells(df_exp, cc_threshold, log_file)\n",
    "\n",
    "    # Protein feature analysis\n",
    "    # df_feat_pro_con, df_result_pro_con = control_group_runner(\n",
    "    #     df_control, pq_writer=writer, log_file=log_file, protein=True\n",
    "    # )\n",
    "    df_feat_pro_exp, df_result_pro_exp = experimental_runner(\n",
    "        df_exp, pq_writer=writer, log_file=log_file, protein=True\n",
    "    )\n",
    "\n",
    "    # # Non-protein feature analysis\n",
    "    # df_feat_no_pro_con, df_result_no_pro_con = control_group_runner(\n",
    "    #     df_control, pq_writer=writer, log_file=log_file, protein=False\n",
    "    # )\n",
    "    # df_feat_no_pro_exp, df_result_no_pro_exp = experimental_runner(\n",
    "    #     df_exp, pq_writer=writer, log_file=log_file, protein=False\n",
    "    # )\n",
    "    # writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0d717c6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Classifier_ID</th>\n",
       "      <th>Plate</th>\n",
       "      <th>trainsize_0</th>\n",
       "      <th>testsize_0</th>\n",
       "      <th>well_0</th>\n",
       "      <th>allele_0</th>\n",
       "      <th>trainsize_1</th>\n",
       "      <th>testsize_1</th>\n",
       "      <th>well_1</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>Metadata_Control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_A24_M10</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1</td>\n",
       "      <td>1153</td>\n",
       "      <td>392</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1261</td>\n",
       "      <td>448</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T2_A24_M10</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T2</td>\n",
       "      <td>1258</td>\n",
       "      <td>287</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1327</td>\n",
       "      <td>382</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T3_A24_M10</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T3</td>\n",
       "      <td>1022</td>\n",
       "      <td>523</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1249</td>\n",
       "      <td>460</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T4_A24_M10</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T4</td>\n",
       "      <td>1202</td>\n",
       "      <td>343</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1290</td>\n",
       "      <td>419</td>\n",
       "      <td>M10</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_A24_B01</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1</td>\n",
       "      <td>1153</td>\n",
       "      <td>392</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1933</td>\n",
       "      <td>539</td>\n",
       "      <td>B01</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T2_A24_B01</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T2</td>\n",
       "      <td>1258</td>\n",
       "      <td>287</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1851</td>\n",
       "      <td>621</td>\n",
       "      <td>B01</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T3_A24_B01</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T3</td>\n",
       "      <td>1022</td>\n",
       "      <td>523</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1870</td>\n",
       "      <td>602</td>\n",
       "      <td>B01</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T4_A24_B01</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T4</td>\n",
       "      <td>1202</td>\n",
       "      <td>343</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>1762</td>\n",
       "      <td>710</td>\n",
       "      <td>B01</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T1_A24_B02</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T1</td>\n",
       "      <td>1153</td>\n",
       "      <td>392</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>771</td>\n",
       "      <td>252</td>\n",
       "      <td>B02</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T2_A24_B02</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T2</td>\n",
       "      <td>1258</td>\n",
       "      <td>287</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>742</td>\n",
       "      <td>281</td>\n",
       "      <td>B02</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T3_A24_B02</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T3</td>\n",
       "      <td>1022</td>\n",
       "      <td>523</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>785</td>\n",
       "      <td>238</td>\n",
       "      <td>B02</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2025_01_27_B13A7A8P2_T4_A24_B02</td>\n",
       "      <td>2025_01_27_B13A7A8P2_T4</td>\n",
       "      <td>1202</td>\n",
       "      <td>343</td>\n",
       "      <td>A24</td>\n",
       "      <td>CCM2_Ala123Val</td>\n",
       "      <td>771</td>\n",
       "      <td>252</td>\n",
       "      <td>B02</td>\n",
       "      <td>CCM2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Classifier_ID                    Plate  trainsize_0  \\\n",
       "0   2025_01_27_B13A7A8P2_T1_A24_M10  2025_01_27_B13A7A8P2_T1         1153   \n",
       "1   2025_01_27_B13A7A8P2_T2_A24_M10  2025_01_27_B13A7A8P2_T2         1258   \n",
       "2   2025_01_27_B13A7A8P2_T3_A24_M10  2025_01_27_B13A7A8P2_T3         1022   \n",
       "3   2025_01_27_B13A7A8P2_T4_A24_M10  2025_01_27_B13A7A8P2_T4         1202   \n",
       "4   2025_01_27_B13A7A8P2_T1_A24_B01  2025_01_27_B13A7A8P2_T1         1153   \n",
       "5   2025_01_27_B13A7A8P2_T2_A24_B01  2025_01_27_B13A7A8P2_T2         1258   \n",
       "6   2025_01_27_B13A7A8P2_T3_A24_B01  2025_01_27_B13A7A8P2_T3         1022   \n",
       "7   2025_01_27_B13A7A8P2_T4_A24_B01  2025_01_27_B13A7A8P2_T4         1202   \n",
       "8   2025_01_27_B13A7A8P2_T1_A24_B02  2025_01_27_B13A7A8P2_T1         1153   \n",
       "9   2025_01_27_B13A7A8P2_T2_A24_B02  2025_01_27_B13A7A8P2_T2         1258   \n",
       "10  2025_01_27_B13A7A8P2_T3_A24_B02  2025_01_27_B13A7A8P2_T3         1022   \n",
       "11  2025_01_27_B13A7A8P2_T4_A24_B02  2025_01_27_B13A7A8P2_T4         1202   \n",
       "\n",
       "    testsize_0 well_0        allele_0  trainsize_1  testsize_1 well_1  \\\n",
       "0          392    A24  CCM2_Ala123Val         1261         448    M10   \n",
       "1          287    A24  CCM2_Ala123Val         1327         382    M10   \n",
       "2          523    A24  CCM2_Ala123Val         1249         460    M10   \n",
       "3          343    A24  CCM2_Ala123Val         1290         419    M10   \n",
       "4          392    A24  CCM2_Ala123Val         1933         539    B01   \n",
       "5          287    A24  CCM2_Ala123Val         1851         621    B01   \n",
       "6          523    A24  CCM2_Ala123Val         1870         602    B01   \n",
       "7          343    A24  CCM2_Ala123Val         1762         710    B01   \n",
       "8          392    A24  CCM2_Ala123Val          771         252    B02   \n",
       "9          287    A24  CCM2_Ala123Val          742         281    B02   \n",
       "10         523    A24  CCM2_Ala123Val          785         238    B02   \n",
       "11         343    A24  CCM2_Ala123Val          771         252    B02   \n",
       "\n",
       "   allele_1  Metadata_Control  \n",
       "0      CCM2             False  \n",
       "1      CCM2             False  \n",
       "2      CCM2             False  \n",
       "3      CCM2             False  \n",
       "4      CCM2             False  \n",
       "5      CCM2             False  \n",
       "6      CCM2             False  \n",
       "7      CCM2             False  \n",
       "8      CCM2             False  \n",
       "9      CCM2             False  \n",
       "10     CCM2             False  \n",
       "11     CCM2             False  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result_pro_exp"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varchamp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
