{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0e1f9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import sqlite3\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "import black\n",
    "import jupyter_black\n",
    "\n",
    "jupyter_black.load(\n",
    "    lab=False,\n",
    "    line_length=79,\n",
    "    verbosity=\"DEBUG\",\n",
    "    target_version=black.TargetVersion.PY310,\n",
    ")\n",
    "\n",
    "# ignore mix type warnings from pandas\n",
    "import warnings\n",
    "\n",
    "from pycytominer.cyto_utils.cells import SingleCells\n",
    "from pycytominer import annotate, normalize, feature_select\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e97a994c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting file paths\n",
    "data_dir = pathlib.Path(\"../../\").resolve(strict=True)\n",
    "metadata_dir = pathlib.Path(\n",
    "    \"../../2021_09_01_VarChAMP-data/metadata/reprocessed/\"\n",
    ").resolve(strict=True)\n",
    "out_dir = pathlib.Path(\n",
    "    \"../../2021_09_01_VarChAMP_local/6.downstream_analysis/results\"\n",
    ")\n",
    "out_dir.mkdir(exist_ok=True)\n",
    "\n",
    "# input file paths\n",
    "plate = \"2023-05-30_B1A1R1_P1T1\"\n",
    "plate_data = pathlib.Path(data_dir / \"2023-05-30_B1A1R1_P1T1.sqlite\").resolve(\n",
    "    strict=True\n",
    ")\n",
    "plate_map = (metadata_dir / \"2023_05_30_B1A1R1.csv\").resolve(strict=True)\n",
    "\n",
    "# setting output paths\n",
    "sc_profiles_path = out_dir / \"B1A1R1_P1T1_single_cell_profile.csv.gz\"\n",
    "sc_profiles_norm_path = (\n",
    "    out_dir / \"B1A1R1_P1T1_single_cell_profile_normalized.csv.gz\"\n",
    ")\n",
    "sc_profiles_norm_select_path = (\n",
    "    out_dir\n",
    "    / \"B1A1R1_P1T1_single_cell_profile_normalized_feature_selected.csv.gz\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0009f439",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_feature_col(col):\n",
    "    \"\"\"Check if column is a feature\"\"\"\n",
    "    return (\n",
    "        col.startswith(\"Cell\")\n",
    "        or col.startswith(\"Cytoplasm\")\n",
    "        or col.startswith(\"Nuclei\")\n",
    "    )\n",
    "\n",
    "\n",
    "def count(cur, table):\n",
    "    \"\"\"Count total number of rows for a table\"\"\"\n",
    "    (num_rows,) = next(cur.execute(f\"SELECT COUNT(*) FROM {table}\"))\n",
    "    return num_rows\n",
    "\n",
    "\n",
    "def get_columns(cur, table):\n",
    "    \"\"\"Get feature and metadata columns lists\"\"\"\n",
    "    ptr = cur.execute(f\"SELECT * FROM {table} LIMIT 1\")\n",
    "    col_names = [obj[0] for obj in ptr.description]\n",
    "\n",
    "    feat_cols = []\n",
    "    meta_cols = []\n",
    "    for col in col_names:\n",
    "        if is_feature_col(col):\n",
    "            feat_cols.append(col)\n",
    "        else:\n",
    "            meta_cols.append(col)\n",
    "\n",
    "    return meta_cols, feat_cols\n",
    "\n",
    "\n",
    "def load_compartment(cur, table):\n",
    "    # Get data useful to pre-alloc memory\n",
    "    num_cells = count(cur, table)\n",
    "    meta_cols, feat_cols = get_columns(cur, table)\n",
    "    num_meta, num_feats = len(meta_cols), len(feat_cols)\n",
    "\n",
    "    feats = np.empty(shape=(num_cells, num_feats), dtype=np.float32)\n",
    "    meta = pd.DataFrame(columns=meta_cols, index=range(num_cells))\n",
    "\n",
    "    columns = \", \".join(meta_cols + feat_cols)\n",
    "    query = f\"SELECT {columns} from {table}\"\n",
    "    resultset = cur.execute(query)\n",
    "    for i, row in tqdm(enumerate(resultset), total=num_cells):\n",
    "        meta.loc[i] = row[:num_meta]\n",
    "        feats[i] = row[num_meta:]\n",
    "    return pd.concat(\n",
    "        [meta, pd.DataFrame(columns=feat_cols, data=feats)], axis=1\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "214536ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# update compartment names and strata\n",
    "strata = [\"Metadata_Well\", \"Metadata_Plate\"]\n",
    "compartments = [\"Cells\", \"Nuclei\", \"Cytoplasm\"]\n",
    "\n",
    "# Updating linking columns for merging all compartments\n",
    "linking_cols = {\n",
    "    \"Cytoplasm\": {\n",
    "        \"Cells\": \"Cytoplasm_Parent_Cells\",\n",
    "        \"Nuclei\": \"Cytoplasm_Parent_Nuclei\",\n",
    "    },\n",
    "    \"Cells\": {\"Cytoplasm\": \"Cells_Number_Object_Number\"},\n",
    "    \"Nuclei\": {\"Cytoplasm\": \"Nuclei_Number_Object_Number\"},\n",
    "}\n",
    "\n",
    "merge_cols = [\"ImageNumber\", \"TableNumber\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f16b99e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect(plate_data)\n",
    "cur = con.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aaab450",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_df = \"\"\n",
    "linking_check_cols = []\n",
    "merge_suffix_rename = []\n",
    "\n",
    "for left_compartment in linking_cols:\n",
    "    for right_compartment in linking_cols[left_compartment]:\n",
    "        # Make sure only one merge per combination occurs\n",
    "        linking_check = \"-\".join(sorted([left_compartment, right_compartment]))\n",
    "        if linking_check in linking_check_cols:\n",
    "            continue\n",
    "\n",
    "        # Specify how to indicate merge suffixes\n",
    "        merge_suffix = [\n",
    "            \"_{comp_l}\".format(comp_l=left_compartment),\n",
    "            \"_{comp_r}\".format(comp_r=right_compartment),\n",
    "        ]\n",
    "        merge_suffix_rename += merge_suffix\n",
    "        left_link_col = linking_cols[left_compartment][right_compartment]\n",
    "        right_link_col = linking_cols[right_compartment][left_compartment]\n",
    "\n",
    "        if isinstance(sc_df, str):\n",
    "            sc_df = load_compartment(cur, left_compartment)\n",
    "\n",
    "        sc_df = sc_df.merge(\n",
    "            load_compartment(cur, right_compartment),\n",
    "            left_on=merge_cols + [left_link_col],\n",
    "            right_on=merge_cols + [right_link_col],\n",
    "            suffixes=merge_suffix,\n",
    "        )\n",
    "\n",
    "        linking_check_cols.append(linking_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaa80eec",
   "metadata": {},
   "outputs": [],
   "source": [
    "rowdict = {\n",
    "    \"01\": \"A\",\n",
    "    \"02\": \"B\",\n",
    "    \"03\": \"C\",\n",
    "    \"04\": \"D\",\n",
    "    \"05\": \"E\",\n",
    "    \"06\": \"F\",\n",
    "    \"07\": \"G\",\n",
    "    \"08\": \"H\",\n",
    "    \"09\": \"I\",\n",
    "    \"10\": \"J\",\n",
    "    \"11\": \"K\",\n",
    "    \"12\": \"L\",\n",
    "    \"13\": \"M\",\n",
    "    \"14\": \"N\",\n",
    "    \"15\": \"O\",\n",
    "    \"16\": \"P\",\n",
    "}\n",
    "\n",
    "image_df = load_compartment(cur, \"Image\")\n",
    "image_df = image_df[[\"TableNumber\", \"FileName_OrigAGP\"]]\n",
    "image_df[\"Metadata_Plate\"] = plate\n",
    "image_df[\"Metadata_Well\"] = image_df[\"FileName_OrigAGP\"].apply(\n",
    "    lambda x: rowdict[x[1:3]] + x[4:6]\n",
    ")\n",
    "aligned_df = sc_df.merge(image_df, on=[\"TableNumber\"], how=\"left\")\n",
    "aligned_df.drop(\n",
    "    [\n",
    "        \"FileName_OrigAGP\",\n",
    "        \"Cytoplasm_Parent_Cells\",\n",
    "        \"Cytoplasm_Parent_Nuclei\",\n",
    "        \"Cells_Number_Object_Number\",\n",
    "        \"Nuclei_Number_Object_Number\",\n",
    "        \"ObjectNumber_Cytoplasm\",\n",
    "        \"ObjectNumber\",\n",
    "        \"ObjectNumber_Cells\",\n",
    "        \"TableNumber\",\n",
    "        \"ImageNumber\",\n",
    "    ],\n",
    "    inplace=True,\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "meta_df = pd.read_csv(plate_map)\n",
    "meta_df = meta_df[\n",
    "    [\n",
    "        \"Metadata_Plate\",\n",
    "        \"Metadata_Well\",\n",
    "        \"node_type\",\n",
    "        \"allele\",\n",
    "        \"batch\",\n",
    "        \"control\",\n",
    "        \"Gene\",\n",
    "        \"MT\",\n",
    "        \"Variant\",\n",
    "        \"Metadata_Sample_Unique\",\n",
    "        \"Metadata_batch_Plate\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "col_name = []\n",
    "for i in meta_df.columns:\n",
    "    if \"Metadata\" in i:\n",
    "        col_name.append(i)\n",
    "    else:\n",
    "        col_name.append(\"Metadata_\" + i)\n",
    "\n",
    "meta_df = meta_df.set_axis(col_name, axis=1)\n",
    "\n",
    "aligned_df = meta_df.merge(\n",
    "    aligned_df, on=[\"Metadata_Plate\", \"Metadata_Well\"], how=\"right\"\n",
    ")\n",
    "aligned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7db74bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_df.to_csv(sc_profiles_path, index=False, compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c91fab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sql_query = \"\"\"SELECT name FROM sqlite_master  \n",
    "#   WHERE type='table';\"\"\"\n",
    "# (num_rows,) = next(cur.execute('''SELECT COUNT(*) FROM Image'''))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "739cd11e",
   "metadata": {},
   "source": [
    "### Normalization and Feature Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70129dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Whole plate normalization\n",
    "\n",
    "normalize(\n",
    "    profiles=sc_profiles_path,\n",
    "    features=\"infer\",\n",
    "    image_features=False,\n",
    "    meta_features=\"infer\",\n",
    "    samples=\"all\",\n",
    "    method='mad_robustize',\n",
    "    output_file=sc_profiles_norm_path,\n",
    "    compression_options=\"gzip\",\n",
    ")\n",
    "\n",
    "print(f\"Normalized profile saved in: {sc_profiles_norm_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac48cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform plate level feature selection for now\n",
    "feature_select(\n",
    "    profiles=sc_profiles_norm_path,\n",
    "    features=\"infer\",\n",
    "    image_features=True,\n",
    "    samples=\"all\",\n",
    "    operation=[\"variance_threshold\", \"correlation_threshold\", \"blocklist\", \"drop_na_columns\"],\n",
    "    output_file=sc_profiles_norm_select_path,\n",
    "    compression_options=\"gzip\",\n",
    ")\n",
    "\n",
    "print(f\"Feature selected profiles saved in: {sc_profiles_norm_select_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
